{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DSS-AZUL-INDEXING\n",
    "\n",
    "### This jupyter notebook is to do a couple testing indexing operations using the DataExtractor class and the FileIndexer class."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below here we import our modules and set up:\n",
    "* ElasticSearch Client\n",
    "* Dummy payload of event\n",
    "* Parse the bundle_uuid and the bundle_version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from elasticsearch import Elasticsearch\n",
    "from chalicelib.utils import DataExtractor\n",
    "from chalicelib.indexer import FileIndexer, AssayOrientedIndexer\n",
    "import json\n",
    "from pprint import pprint\n",
    "\n",
    "# Create an ElasticSearch client\n",
    "es = Elasticsearch([{'host': 'localhost', 'port': 9200}])\n",
    "# Sample event payload\n",
    "payload = { \"query\": { \"query\": { \"match_all\":{}} }, \"subscription_id\": \"ba50df7b-5a97-4e87-b9ce-c0935a817f0b\", \"transaction_id\": \"ff6b7fa3-dc79-4a79-a313-296801de76b9\", \"match\": { \"bundle_version\": \"2018-01-03T173141.039382Z\", \"bundle_uuid\": \"d8a4576b-66f5-4b10-aecf-f6c4025b9997\" } }\n",
    "bundle_uuid = payload['match']['bundle_uuid']\n",
    "bundle_version = payload['match']['bundle_version']\n",
    "# Create "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will create an instance of the DataExtractor and use it to get the contents from the bundle referenced by the variable `payload`. We will be pulling from the AWS replica."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "#####################################################\n",
      "#                    PRINTING METADATA               #\n",
      "#####################################################\n",
      "{   'assay.json': {   'content': {   'assay_id': 'assay_1',\n",
      "                                     'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/assay.json',\n",
      "                                                 'schema_version': '4.6.1',\n",
      "                                                 'type': 'assay'},\n",
      "                                     'rna': {   'end_bias': 'five_prime_end',\n",
      "                                                'library_construction': 'smart-seq2',\n",
      "                                                'strand': 'both'},\n",
      "                                     'seq': {   'instrument_model': 'HiSeq '\n",
      "                                                                    '2500',\n",
      "                                                'instrument_platform': 'Illumina',\n",
      "                                                'lanes': [   {   'number': 1,\n",
      "                                                                 'r1': 'R1.fastq.gz',\n",
      "                                                                 'r2': 'R2.fastq.gz'}],\n",
      "                                                'molecule': 'polyA RNA',\n",
      "                                                'paired_ends': True},\n",
      "                                     'single_cell': {'cell_handling': 'FACS'}},\n",
      "                      'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/assay_bundle.json',\n",
      "                                  'schema_version': '4.6.1',\n",
      "                                  'type': 'assay_bundle'},\n",
      "                      'has_input': '2e377554-32bb-4ded-abc9-a8ada91798de',\n",
      "                      'has_output': [   'a3aaa566-6724-4f47-accb-e57769a077e4',\n",
      "                                        '5da5ace1-e8ca-4454-9d8f-d0105b984632'],\n",
      "                      'hca_ingest': {   'accession': '',\n",
      "                                        'document_id': '57436d7a-aa39-4bfa-82fb-5fdcdaf2d799',\n",
      "                                        'lastModifiedUser': 'anonymousUser',\n",
      "                                        'submissionDate': '2018-01-03T17:16:35.003Z',\n",
      "                                        'updateDate': '2018-01-03T17:17:15.804Z',\n",
      "                                        'user': 'anonymousUser'}},\n",
      "    'project.json': {   'content': {   'contributors': [   {   'email': 'mfreeberg@ebi.ac.uk',\n",
      "                                                               'name': 'Mallory '\n",
      "                                                                       'Freeberg'}],\n",
      "                                       'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project.json',\n",
      "                                                   'type': 'project'},\n",
      "                                       'description': 'Q4_DEMO-We report '\n",
      "                                                      'transcriptomes from 430 '\n",
      "                                                      'single glioblastoma '\n",
      "                                                      'cells isolated from 5 '\n",
      "                                                      'individual tumors and '\n",
      "                                                      '102 single cells from '\n",
      "                                                      'gliomasphere cells '\n",
      "                                                      'lines generated using '\n",
      "                                                      'SMART-seq. In addition, '\n",
      "                                                      'we report population '\n",
      "                                                      'RNA-seq from the five '\n",
      "                                                      'tumors as well as '\n",
      "                                                      'RNA-seq from cell lines '\n",
      "                                                      'derived from 3 tumors '\n",
      "                                                      '(MGH26, MGH28, MGH31) '\n",
      "                                                      'cultured under serum '\n",
      "                                                      'free (GSC) and '\n",
      "                                                      'differentiated (DGC) '\n",
      "                                                      'conditions. This '\n",
      "                                                      'dataset highlights '\n",
      "                                                      'intratumoral '\n",
      "                                                      'heterogeneity with '\n",
      "                                                      'regards to the '\n",
      "                                                      'expression of de novo '\n",
      "                                                      'derived transcriptional '\n",
      "                                                      'modules and established '\n",
      "                                                      'subtype classifiers. '\n",
      "                                                      'Overall design: '\n",
      "                                                      'Operative specimens '\n",
      "                                                      'from five glioblastoma '\n",
      "                                                      'patients (MGH26, MGH28, '\n",
      "                                                      'MGH29, MGH30, MGH31) '\n",
      "                                                      'were acutely '\n",
      "                                                      'dissociated, depleted '\n",
      "                                                      'for CD45+ inflammatory '\n",
      "                                                      'cells and then sorted '\n",
      "                                                      'as single cells (576 '\n",
      "                                                      'samples). Population '\n",
      "                                                      'controls for each tumor '\n",
      "                                                      'were isolated by '\n",
      "                                                      'sorting 2000-10000 '\n",
      "                                                      'cells and processed in '\n",
      "                                                      'parallel (5 population '\n",
      "                                                      'control samples). '\n",
      "                                                      'Single cells from two '\n",
      "                                                      'established cell lines, '\n",
      "                                                      'GBM6 and GBM8, were '\n",
      "                                                      'also sorted as single '\n",
      "                                                      'cells (192 samples). '\n",
      "                                                      'SMART-seq protocol was '\n",
      "                                                      'implemented to generate '\n",
      "                                                      'single cell full length '\n",
      "                                                      'transcriptomes '\n",
      "                                                      '(modified from Shalek, '\n",
      "                                                      'et al Nature 2013) and '\n",
      "                                                      'sequenced using 25 bp '\n",
      "                                                      'paired end reads. '\n",
      "                                                      'Single cell cDNA '\n",
      "                                                      'libraries for MGH30 '\n",
      "                                                      'were resequenced using '\n",
      "                                                      '100 bp paired end reads '\n",
      "                                                      'to allow for isoform '\n",
      "                                                      'and splice junction '\n",
      "                                                      'reconstruction (96 '\n",
      "                                                      'samples, annotated '\n",
      "                                                      'MGH30L). Cells were '\n",
      "                                                      'also cultured in serum '\n",
      "                                                      'free conditions to '\n",
      "                                                      'generate gliomasphere '\n",
      "                                                      'cell lines for MGH26, '\n",
      "                                                      'MGH28, and MGH31 (GSC) '\n",
      "                                                      'which were then '\n",
      "                                                      'differentiated using '\n",
      "                                                      '10% serum (DGC). '\n",
      "                                                      'Population RNA-seq was '\n",
      "                                                      'performed on these '\n",
      "                                                      'samples (3 GSC, 3 DGC, '\n",
      "                                                      '6 total). The initial '\n",
      "                                                      'dataset included 875 '\n",
      "                                                      'RNA-seq libraries (576 '\n",
      "                                                      'single glioblastoma '\n",
      "                                                      'cells, 96 resequenced '\n",
      "                                                      'MGH30L, 192 single '\n",
      "                                                      'gliomasphere cells, 5 '\n",
      "                                                      'tumor population '\n",
      "                                                      'controls, 6 population '\n",
      "                                                      'libraries from GSC and '\n",
      "                                                      'DGC samples). Data was '\n",
      "                                                      'processed as described '\n",
      "                                                      'below using RSEM for '\n",
      "                                                      'quantification of gene '\n",
      "                                                      'expression. 5,948 genes '\n",
      "                                                      'with the highest '\n",
      "                                                      'composite expression '\n",
      "                                                      'either across all '\n",
      "                                                      'single cells combined '\n",
      "                                                      '(average log2(TPM)>4.5) '\n",
      "                                                      'or within a single '\n",
      "                                                      'tumor (average '\n",
      "                                                      'log2(TPM)>6 in at least '\n",
      "                                                      'one tumor) were '\n",
      "                                                      'included. Cells '\n",
      "                                                      'expressing less than '\n",
      "                                                      '2,000 of these 5,948 '\n",
      "                                                      'genes were excluded. '\n",
      "                                                      'The final processed '\n",
      "                                                      'dataset then included '\n",
      "                                                      '430 primary single cell '\n",
      "                                                      'glioblastoma '\n",
      "                                                      'transcriptomes, 102 '\n",
      "                                                      'single cell '\n",
      "                                                      'transcriptomes from '\n",
      "                                                      'cell lines(GBM6,GBM8), '\n",
      "                                                      '5 population controls '\n",
      "                                                      '(1 for each tumor), and '\n",
      "                                                      '6 population libraries '\n",
      "                                                      'from cell lines derived '\n",
      "                                                      'from the tumors (GSC '\n",
      "                                                      'and DGC for MGH26, '\n",
      "                                                      'MGH28 and MGH31). The '\n",
      "                                                      'final matrix '\n",
      "                                                      '(GBM_data_matrix.txt) '\n",
      "                                                      'therefore contains 5948 '\n",
      "                                                      'rows (genes) quantified '\n",
      "                                                      'in 543 samples '\n",
      "                                                      '(columns). Please note '\n",
      "                                                      'that the samples which '\n",
      "                                                      'are not included in the '\n",
      "                                                      'data processing are '\n",
      "                                                      'indicated in the sample '\n",
      "                                                      'description field.\\n',\n",
      "                                       'name': 'Q4_DEMO-Single cell RNA-seq of '\n",
      "                                               'primary human glioblastomas',\n",
      "                                       'project_id': 'Q4_DEMO-project_PRJNA248302'},\n",
      "                        'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project_bundle.json',\n",
      "                                    'schema_version': '4.6.1',\n",
      "                                    'type': 'project_bundle'},\n",
      "                        'hca_ingest': {   'accession': '',\n",
      "                                          'document_id': '617ae8e2-a994-4451-a81e-0d5f652ed4da',\n",
      "                                          'lastModifiedUser': 'anonymousUser',\n",
      "                                          'submissionDate': '2018-01-03T17:13:44.555Z',\n",
      "                                          'updateDate': '2018-01-03T17:16:09.887Z',\n",
      "                                          'user': 'google-oauth2|111466450643076624860'}},\n",
      "    'sample.json': {   'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample_bundle.json',\n",
      "                                   'schema_version': '4.6.1',\n",
      "                                   'type': 'sample_bundle'},\n",
      "                       'samples': [   {   'content': {   'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                                                                     'schema_version': '4.6.1',\n",
      "                                                                     'type': 'sample'},\n",
      "                                                         'derived_from': 'Q4_DEMO-donor_MGH30',\n",
      "                                                         'genus_species': {   'ontology': 'NCBITaxon:9606',\n",
      "                                                                              'text': 'Homo '\n",
      "                                                                                      'sapiens'},\n",
      "                                                         'name': 'Q4_DEMO-Single '\n",
      "                                                                 'cell '\n",
      "                                                                 'mRNA-seq_MGH30_A01',\n",
      "                                                         'ncbi_taxon_id': 9606,\n",
      "                                                         'sample_id': 'Q4_DEMO-sample_SAMN02797092',\n",
      "                                                         'specimen_from_organism': {   'body_part': {   'text': 'glioblastoma'},\n",
      "                                                                                       'organ': {   'ontology': 'UBERON:0000955',\n",
      "                                                                                                    'text': 'brain'}}},\n",
      "                                          'derivation_protocols': [   {   'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/protocol.json',\n",
      "                                                                                      'schema_version': '4.6.1',\n",
      "                                                                                      'type': 'protocol'},\n",
      "                                                                          'pdf': 'Q4_DEMO-protocol.pdf',\n",
      "                                                                          'protocol_id': 'Q4_DEMO-protocol',\n",
      "                                                                          'type': {   'text': 'single '\n",
      "                                                                                              'cell '\n",
      "                                                                                              'sequencing'}}],\n",
      "                                          'derived_from': '176ff29c-e492-495b-a537-d21940a53799',\n",
      "                                          'hca_ingest': {   'accession': '',\n",
      "                                                            'document_id': '2e377554-32bb-4ded-abc9-a8ada91798de',\n",
      "                                                            'lastModifiedUser': 'anonymousUser',\n",
      "                                                            'submissionDate': '2018-01-03T17:16:34.953Z',\n",
      "                                                            'updateDate': '2018-01-03T17:17:17.116Z',\n",
      "                                                            'user': 'anonymousUser'}},\n",
      "                                      {   'content': {   'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                                                                     'schema_version': '4.6.1',\n",
      "                                                                     'type': 'sample'},\n",
      "                                                         'donor': {   'is_living': True},\n",
      "                                                         'genus_species': {   'ontology': 'NCBITaxon:9606',\n",
      "                                                                              'text': 'Homo '\n",
      "                                                                                      'sapiens'},\n",
      "                                                         'name': 'Q4 DEMO '\n",
      "                                                                 'donor MGH30',\n",
      "                                                         'ncbi_taxon_id': 9606,\n",
      "                                                         'sample_id': 'Q4_DEMO-donor_MGH30'},\n",
      "                                          'derivation_protocols': [],\n",
      "                                          'hca_ingest': {   'accession': '',\n",
      "                                                            'document_id': '176ff29c-e492-495b-a537-d21940a53799',\n",
      "                                                            'lastModifiedUser': 'anonymousUser',\n",
      "                                                            'submissionDate': '2018-01-03T17:16:34.938Z',\n",
      "                                                            'updateDate': '2018-01-03T17:17:15.915Z',\n",
      "                                                            'user': 'anonymousUser'}}]}}\n",
      "\n",
      "#####################################################\n",
      "#                  PRINTING DATA FILES               #\n",
      "#####################################################\n",
      "{   'R1.fastq.gz': {   'content-type': 'application/gzip; dcp-type=data',\n",
      "                       'crc32c': '4ef74578',\n",
      "                       'indexed': False,\n",
      "                       'name': 'R1.fastq.gz',\n",
      "                       's3_etag': 'c7bbee4c46bbf29432862e05830c8f39',\n",
      "                       'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3',\n",
      "                       'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29',\n",
      "                       'size': 125191,\n",
      "                       'uuid': '5da5ace1-e8ca-4454-9d8f-d0105b984632',\n",
      "                       'version': '2018-01-03T173137.798791Z'},\n",
      "    'R2.fastq.gz': {   'content-type': 'application/gzip; dcp-type=data',\n",
      "                       'crc32c': '69987b3e',\n",
      "                       'indexed': False,\n",
      "                       'name': 'R2.fastq.gz',\n",
      "                       's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae',\n",
      "                       'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21',\n",
      "                       'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba',\n",
      "                       'size': 130024,\n",
      "                       'uuid': 'a3aaa566-6724-4f47-accb-e57769a077e4',\n",
      "                       'version': '2018-01-03T173138.944258Z'}}\n"
     ]
    }
   ],
   "source": [
    "# Create DataExtractor instance pointing to HCA Staging\n",
    "extractor = DataExtractor(\"https://dss.staging.data.humancellatlas.org/v1\")\n",
    "# Use dummy payload and get the metadata_files and the data_files\n",
    "metadata_files, data_files = extractor.extract_bundle(payload, \"aws\")\n",
    "# Print each dictionary\n",
    "print(\"\\n#####################################################\")\n",
    "print(\"#                    PRINTING METADATA               #\")\n",
    "print(\"#####################################################\")\n",
    "pprint(metadata_files, indent=4)\n",
    "print(\"\\n#####################################################\")\n",
    "print(\"#                  PRINTING DATA FILES               #\")\n",
    "print(\"#####################################################\")\n",
    "pprint(data_files, indent=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we pass this on to the FileIndexer class to create a File Oriented index entry on ElasticSearch running on `localhost:9200`. But first, we get the index settings and get the configuration files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Content: assay_id ; Name: assayId \n",
      "\n",
      "(\"THE LEAF _file is: {'single_cell': {'cell_handling': 'FACS'}, 'core': \"\n",
      " \"{'type': 'assay', 'schema_url': \"\n",
      " \"'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/assay.json', \"\n",
      " \"'schema_version': '4.6.1'}, 'rna': {'end_bias': 'five_prime_end', 'strand': \"\n",
      " \"'both', 'library_construction': 'smart-seq2'}, 'assay_id': 'assay_1', 'seq': \"\n",
      " \"{'instrument_platform': 'Illumina', 'molecule': 'polyA RNA', 'paired_ends': \"\n",
      " \"True, 'lanes': [{'number': 1, 'r2': 'R2.fastq.gz', 'r1': 'R1.fastq.gz'}], \"\n",
      " \"'instrument_model': 'HiSeq 2500'}}\\n\"\n",
      " ' ')\n",
      "the file content is: assay_1\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>, {'assayId': ['assay_1']})\n",
      "Content: cell_handling ; Name: scCellHandling \n",
      "\n",
      "\"THE LEAF _file is: {'cell_handling': 'FACS'}\\n \"\n",
      "the file content is: FACS\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'scCellHandling': ['FACS']})\n",
      "Content: instrument_platform ; Name: seqInstrumentPlatform \n",
      "\n",
      "(\"THE LEAF _file is: {'instrument_platform': 'Illumina', 'molecule': 'polyA \"\n",
      " \"RNA', 'paired_ends': True, 'lanes': [{'number': 1, 'r2': 'R2.fastq.gz', \"\n",
      " \"'r1': 'R1.fastq.gz'}], 'instrument_model': 'HiSeq 2500'}\\n\"\n",
      " ' ')\n",
      "the file content is: Illumina\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "Content: paired_ends ; Name: pairedEnds \n",
      "\n",
      "(\"THE LEAF _file is: {'instrument_platform': 'Illumina', 'molecule': 'polyA \"\n",
      " \"RNA', 'paired_ends': True, 'lanes': [{'number': 1, 'r2': 'R2.fastq.gz', \"\n",
      " \"'r1': 'R1.fastq.gz'}], 'instrument_model': 'HiSeq 2500'}\\n\"\n",
      " ' ')\n",
      "the file content is: True\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "Content: library_construction ; Name: rnaLibraryConstruction \n",
      "\n",
      "(\"THE LEAF _file is: {'end_bias': 'five_prime_end', 'strand': 'both', \"\n",
      " \"'library_construction': 'smart-seq2'}\\n\"\n",
      " ' ')\n",
      "the file content is: smart-seq2\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "Content: spike_in ; Name: rnaSpikeIn \n",
      "\n",
      "(\"THE LEAF _file is: {'end_bias': 'five_prime_end', 'strand': 'both', \"\n",
      " \"'library_construction': 'smart-seq2'}\\n\"\n",
      " ' ')\n",
      "Content: project_id ; Name: projectId \n",
      "\n",
      "(\"THE LEAF _file is: {'core': {'type': 'project', 'schema_url': \"\n",
      " \"'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project.json'}, \"\n",
      " \"'project_id': 'Q4_DEMO-project_PRJNA248302', 'name': 'Q4_DEMO-Single cell \"\n",
      " \"RNA-seq of primary human glioblastomas', 'contributors': [{'email': \"\n",
      " \"'mfreeberg@ebi.ac.uk', 'name': 'Mallory Freeberg'}], 'description': \"\n",
      " \"'Q4_DEMO-We report transcriptomes from 430 single glioblastoma cells \"\n",
      " 'isolated from 5 individual tumors and 102 single cells from gliomasphere '\n",
      " 'cells lines generated using SMART-seq. In addition, we report population '\n",
      " 'RNA-seq from the five tumors as well as RNA-seq from cell lines derived from '\n",
      " '3 tumors (MGH26, MGH28, MGH31) cultured under serum free (GSC) and '\n",
      " 'differentiated (DGC) conditions. This dataset highlights intratumoral '\n",
      " 'heterogeneity with regards to the expression of de novo derived '\n",
      " 'transcriptional modules and established subtype classifiers. Overall design: '\n",
      " 'Operative specimens from five glioblastoma patients (MGH26, MGH28, MGH29, '\n",
      " 'MGH30, MGH31) were acutely dissociated, depleted for CD45+ inflammatory '\n",
      " 'cells and then sorted as single cells (576 samples). Population controls for '\n",
      " 'each tumor were isolated by sorting 2000-10000 cells and processed in '\n",
      " 'parallel (5 population control samples). Single cells from two established '\n",
      " 'cell lines, GBM6 and GBM8, were also sorted as single cells (192 samples). '\n",
      " 'SMART-seq protocol was implemented to generate single cell full length '\n",
      " 'transcriptomes (modified from Shalek, et al Nature 2013) and sequenced using '\n",
      " '25 bp paired end reads. Single cell cDNA libraries for MGH30 were '\n",
      " 'resequenced using 100 bp paired end reads to allow for isoform and splice '\n",
      " 'junction reconstruction (96 samples, annotated MGH30L). Cells were also '\n",
      " 'cultured in serum free conditions to generate gliomasphere cell lines for '\n",
      " 'MGH26, MGH28, and MGH31 (GSC) which were then differentiated using 10% serum '\n",
      " '(DGC). Population RNA-seq was performed on these samples (3 GSC, 3 DGC, 6 '\n",
      " 'total). The initial dataset included 875 RNA-seq libraries (576 single '\n",
      " 'glioblastoma cells, 96 resequenced MGH30L, 192 single gliomasphere cells, 5 '\n",
      " 'tumor population controls, 6 population libraries from GSC and DGC samples). '\n",
      " 'Data was processed as described below using RSEM for quantification of gene '\n",
      " 'expression. 5,948 genes with the highest composite expression either across '\n",
      " 'all single cells combined (average log2(TPM)>4.5) or within a single tumor '\n",
      " '(average log2(TPM)>6 in at least one tumor) were included. Cells expressing '\n",
      " 'less than 2,000 of these 5,948 genes were excluded. The final processed '\n",
      " 'dataset then included 430 primary single cell glioblastoma transcriptomes, '\n",
      " '102 single cell transcriptomes from cell lines(GBM6,GBM8), 5 population '\n",
      " 'controls (1 for each tumor), and 6 population libraries from cell lines '\n",
      " 'derived from the tumors (GSC and DGC for MGH26, MGH28 and MGH31). The final '\n",
      " 'matrix (GBM_data_matrix.txt) therefore contains 5948 rows (genes) quantified '\n",
      " 'in 543 samples (columns). Please note that the samples which are not '\n",
      " 'included in the data processing are indicated in the sample description '\n",
      " \"field.\\\\n'}\\n\"\n",
      " ' ')\n",
      "the file content is: Q4_DEMO-project_PRJNA248302\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "Content: submitters ; Name: projectSubmitters \n",
      "\n",
      "(\"THE LEAF _file is: {'core': {'type': 'project', 'schema_url': \"\n",
      " \"'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project.json'}, \"\n",
      " \"'project_id': 'Q4_DEMO-project_PRJNA248302', 'name': 'Q4_DEMO-Single cell \"\n",
      " \"RNA-seq of primary human glioblastomas', 'contributors': [{'email': \"\n",
      " \"'mfreeberg@ebi.ac.uk', 'name': 'Mallory Freeberg'}], 'description': \"\n",
      " \"'Q4_DEMO-We report transcriptomes from 430 single glioblastoma cells \"\n",
      " 'isolated from 5 individual tumors and 102 single cells from gliomasphere '\n",
      " 'cells lines generated using SMART-seq. In addition, we report population '\n",
      " 'RNA-seq from the five tumors as well as RNA-seq from cell lines derived from '\n",
      " '3 tumors (MGH26, MGH28, MGH31) cultured under serum free (GSC) and '\n",
      " 'differentiated (DGC) conditions. This dataset highlights intratumoral '\n",
      " 'heterogeneity with regards to the expression of de novo derived '\n",
      " 'transcriptional modules and established subtype classifiers. Overall design: '\n",
      " 'Operative specimens from five glioblastoma patients (MGH26, MGH28, MGH29, '\n",
      " 'MGH30, MGH31) were acutely dissociated, depleted for CD45+ inflammatory '\n",
      " 'cells and then sorted as single cells (576 samples). Population controls for '\n",
      " 'each tumor were isolated by sorting 2000-10000 cells and processed in '\n",
      " 'parallel (5 population control samples). Single cells from two established '\n",
      " 'cell lines, GBM6 and GBM8, were also sorted as single cells (192 samples). '\n",
      " 'SMART-seq protocol was implemented to generate single cell full length '\n",
      " 'transcriptomes (modified from Shalek, et al Nature 2013) and sequenced using '\n",
      " '25 bp paired end reads. Single cell cDNA libraries for MGH30 were '\n",
      " 'resequenced using 100 bp paired end reads to allow for isoform and splice '\n",
      " 'junction reconstruction (96 samples, annotated MGH30L). Cells were also '\n",
      " 'cultured in serum free conditions to generate gliomasphere cell lines for '\n",
      " 'MGH26, MGH28, and MGH31 (GSC) which were then differentiated using 10% serum '\n",
      " '(DGC). Population RNA-seq was performed on these samples (3 GSC, 3 DGC, 6 '\n",
      " 'total). The initial dataset included 875 RNA-seq libraries (576 single '\n",
      " 'glioblastoma cells, 96 resequenced MGH30L, 192 single gliomasphere cells, 5 '\n",
      " 'tumor population controls, 6 population libraries from GSC and DGC samples). '\n",
      " 'Data was processed as described below using RSEM for quantification of gene '\n",
      " 'expression. 5,948 genes with the highest composite expression either across '\n",
      " 'all single cells combined (average log2(TPM)>4.5) or within a single tumor '\n",
      " '(average log2(TPM)>6 in at least one tumor) were included. Cells expressing '\n",
      " 'less than 2,000 of these 5,948 genes were excluded. The final processed '\n",
      " 'dataset then included 430 primary single cell glioblastoma transcriptomes, '\n",
      " '102 single cell transcriptomes from cell lines(GBM6,GBM8), 5 population '\n",
      " 'controls (1 for each tumor), and 6 population libraries from cell lines '\n",
      " 'derived from the tumors (GSC and DGC for MGH26, MGH28 and MGH31). The final '\n",
      " 'matrix (GBM_data_matrix.txt) therefore contains 5948 rows (genes) quantified '\n",
      " 'in 543 samples (columns). Please note that the samples which are not '\n",
      " 'included in the data processing are indicated in the sample description '\n",
      " \"field.\\\\n'}\\n\"\n",
      " ' ')\n",
      "Content: email ; Name: projectContributosEmail \n",
      "\n",
      "(\"THE LEAF _file is: [{'email': 'mfreeberg@ebi.ac.uk', 'name': 'Mallory \"\n",
      " \"Freeberg'}]\\n\"\n",
      " ' ')\n",
      "PRINTING LEAF el\n",
      "{'email': 'mfreeberg@ebi.ac.uk', 'name': 'Mallory Freeberg'}\n",
      "Content: email ; Name: projectContributosEmail \n",
      "\n",
      "(\"THE LEAF _file is: {'email': 'mfreeberg@ebi.ac.uk', 'name': 'Mallory \"\n",
      " \"Freeberg'}\\n\"\n",
      " ' ')\n",
      "the file content is: mfreeberg@ebi.ac.uk\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectContributosEmail': ['mfreeberg@ebi.ac.uk'],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "PRINTING _file:\n",
      "[{'content': {'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                       'schema_version': '4.6.1',\n",
      "                       'type': 'sample'},\n",
      "              'derived_from': 'Q4_DEMO-donor_MGH30',\n",
      "              'genus_species': {'ontology': 'NCBITaxon:9606',\n",
      "                                'text': 'Homo sapiens'},\n",
      "              'name': 'Q4_DEMO-Single cell mRNA-seq_MGH30_A01',\n",
      "              'ncbi_taxon_id': 9606,\n",
      "              'sample_id': 'Q4_DEMO-sample_SAMN02797092',\n",
      "              'specimen_from_organism': {'body_part': {'text': 'glioblastoma'},\n",
      "                                         'organ': {'ontology': 'UBERON:0000955',\n",
      "                                                   'text': 'brain'}}},\n",
      "  'derivation_protocols': [{'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/protocol.json',\n",
      "                                     'schema_version': '4.6.1',\n",
      "                                     'type': 'protocol'},\n",
      "                            'pdf': 'Q4_DEMO-protocol.pdf',\n",
      "                            'protocol_id': 'Q4_DEMO-protocol',\n",
      "                            'type': {'text': 'single cell sequencing'}}],\n",
      "  'derived_from': '176ff29c-e492-495b-a537-d21940a53799',\n",
      "  'hca_ingest': {'accession': '',\n",
      "                 'document_id': '2e377554-32bb-4ded-abc9-a8ada91798de',\n",
      "                 'lastModifiedUser': 'anonymousUser',\n",
      "                 'submissionDate': '2018-01-03T17:16:34.953Z',\n",
      "                 'updateDate': '2018-01-03T17:17:17.116Z',\n",
      "                 'user': 'anonymousUser'}},\n",
      " {'content': {'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                       'schema_version': '4.6.1',\n",
      "                       'type': 'sample'},\n",
      "              'donor': {'is_living': True},\n",
      "              'genus_species': {'ontology': 'NCBITaxon:9606',\n",
      "                                'text': 'Homo sapiens'},\n",
      "              'name': 'Q4 DEMO donor MGH30',\n",
      "              'ncbi_taxon_id': 9606,\n",
      "              'sample_id': 'Q4_DEMO-donor_MGH30'},\n",
      "  'derivation_protocols': [],\n",
      "  'hca_ingest': {'accession': '',\n",
      "                 'document_id': '176ff29c-e492-495b-a537-d21940a53799',\n",
      "                 'lastModifiedUser': 'anonymousUser',\n",
      "                 'submissionDate': '2018-01-03T17:16:34.938Z',\n",
      "                 'updateDate': '2018-01-03T17:17:15.915Z',\n",
      "                 'user': 'anonymousUser'}}]\n",
      "PRINTING el\n",
      "{'content': {'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                      'schema_version': '4.6.1',\n",
      "                      'type': 'sample'},\n",
      "             'derived_from': 'Q4_DEMO-donor_MGH30',\n",
      "             'genus_species': {'ontology': 'NCBITaxon:9606',\n",
      "                               'text': 'Homo sapiens'},\n",
      "             'name': 'Q4_DEMO-Single cell mRNA-seq_MGH30_A01',\n",
      "             'ncbi_taxon_id': 9606,\n",
      "             'sample_id': 'Q4_DEMO-sample_SAMN02797092',\n",
      "             'specimen_from_organism': {'body_part': {'text': 'glioblastoma'},\n",
      "                                        'organ': {'ontology': 'UBERON:0000955',\n",
      "                                                  'text': 'brain'}}},\n",
      " 'derivation_protocols': [{'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/protocol.json',\n",
      "                                    'schema_version': '4.6.1',\n",
      "                                    'type': 'protocol'},\n",
      "                           'pdf': 'Q4_DEMO-protocol.pdf',\n",
      "                           'protocol_id': 'Q4_DEMO-protocol',\n",
      "                           'type': {'text': 'single cell sequencing'}}],\n",
      " 'derived_from': '176ff29c-e492-495b-a537-d21940a53799',\n",
      " 'hca_ingest': {'accession': '',\n",
      "                'document_id': '2e377554-32bb-4ded-abc9-a8ada91798de',\n",
      "                'lastModifiedUser': 'anonymousUser',\n",
      "                'submissionDate': '2018-01-03T17:16:34.953Z',\n",
      "                'updateDate': '2018-01-03T17:17:17.116Z',\n",
      "                'user': 'anonymousUser'}}\n",
      "Content: text ; Name: sampleBodyPart \n",
      "\n",
      "\"THE LEAF _file is: {'text': 'glioblastoma'}\\n \"\n",
      "the file content is: glioblastoma\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectContributosEmail': ['mfreeberg@ebi.ac.uk'],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'sampleBodyPart': ['glioblastoma'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "Content: text ; Name: sampleSpecies \n",
      "\n",
      "\"THE LEAF _file is: {'text': 'Homo sapiens', 'ontology': 'NCBITaxon:9606'}\\n \"\n",
      "the file content is: Homo sapiens\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectContributosEmail': ['mfreeberg@ebi.ac.uk'],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'sampleBodyPart': ['glioblastoma'],\n",
      "             'sampleSpecies': ['Homo sapiens'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "PRINTING el\n",
      "{'content': {'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                      'schema_version': '4.6.1',\n",
      "                      'type': 'sample'},\n",
      "             'donor': {'is_living': True},\n",
      "             'genus_species': {'ontology': 'NCBITaxon:9606',\n",
      "                               'text': 'Homo sapiens'},\n",
      "             'name': 'Q4 DEMO donor MGH30',\n",
      "             'ncbi_taxon_id': 9606,\n",
      "             'sample_id': 'Q4_DEMO-donor_MGH30'},\n",
      " 'derivation_protocols': [],\n",
      " 'hca_ingest': {'accession': '',\n",
      "                'document_id': '176ff29c-e492-495b-a537-d21940a53799',\n",
      "                'lastModifiedUser': 'anonymousUser',\n",
      "                'submissionDate': '2018-01-03T17:16:34.938Z',\n",
      "                'updateDate': '2018-01-03T17:17:15.915Z',\n",
      "                'user': 'anonymousUser'}}\n",
      "Content: text ; Name: sampleSpecies \n",
      "\n",
      "\"THE LEAF _file is: {'text': 'Homo sapiens', 'ontology': 'NCBITaxon:9606'}\\n \"\n",
      "the file content is: Homo sapiens\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectContributosEmail': ['mfreeberg@ebi.ac.uk'],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'sampleBodyPart': ['glioblastoma'],\n",
      "             'sampleSpecies': ['Homo sapiens', 'Homo sapiens'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "#######TESTING NEW get_item2 METHOD##########\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectContributosEmail': ['mfreeberg@ebi.ac.uk'],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'sampleBodyPart': ['glioblastoma'],\n",
      "             'sampleSpecies': ['Homo sapiens', 'Homo sapiens'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "Content: assay_id ; Name: assayId \n",
      "\n",
      "(\"THE LEAF _file is: {'single_cell': {'cell_handling': 'FACS'}, 'core': \"\n",
      " \"{'type': 'assay', 'schema_url': \"\n",
      " \"'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/assay.json', \"\n",
      " \"'schema_version': '4.6.1'}, 'rna': {'end_bias': 'five_prime_end', 'strand': \"\n",
      " \"'both', 'library_construction': 'smart-seq2'}, 'assay_id': 'assay_1', 'seq': \"\n",
      " \"{'instrument_platform': 'Illumina', 'molecule': 'polyA RNA', 'paired_ends': \"\n",
      " \"True, 'lanes': [{'number': 1, 'r2': 'R2.fastq.gz', 'r1': 'R1.fastq.gz'}], \"\n",
      " \"'instrument_model': 'HiSeq 2500'}}\\n\"\n",
      " ' ')\n",
      "the file content is: assay_1\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>, {'assayId': ['assay_1']})\n",
      "Content: cell_handling ; Name: scCellHandling \n",
      "\n",
      "\"THE LEAF _file is: {'cell_handling': 'FACS'}\\n \"\n",
      "the file content is: FACS\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'scCellHandling': ['FACS']})\n",
      "Content: instrument_platform ; Name: seqInstrumentPlatform \n",
      "\n",
      "(\"THE LEAF _file is: {'instrument_platform': 'Illumina', 'molecule': 'polyA \"\n",
      " \"RNA', 'paired_ends': True, 'lanes': [{'number': 1, 'r2': 'R2.fastq.gz', \"\n",
      " \"'r1': 'R1.fastq.gz'}], 'instrument_model': 'HiSeq 2500'}\\n\"\n",
      " ' ')\n",
      "the file content is: Illumina\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "Content: paired_ends ; Name: pairedEnds \n",
      "\n",
      "(\"THE LEAF _file is: {'instrument_platform': 'Illumina', 'molecule': 'polyA \"\n",
      " \"RNA', 'paired_ends': True, 'lanes': [{'number': 1, 'r2': 'R2.fastq.gz', \"\n",
      " \"'r1': 'R1.fastq.gz'}], 'instrument_model': 'HiSeq 2500'}\\n\"\n",
      " ' ')\n",
      "the file content is: True\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "Content: library_construction ; Name: rnaLibraryConstruction \n",
      "\n",
      "(\"THE LEAF _file is: {'end_bias': 'five_prime_end', 'strand': 'both', \"\n",
      " \"'library_construction': 'smart-seq2'}\\n\"\n",
      " ' ')\n",
      "the file content is: smart-seq2\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "Content: spike_in ; Name: rnaSpikeIn \n",
      "\n",
      "(\"THE LEAF _file is: {'end_bias': 'five_prime_end', 'strand': 'both', \"\n",
      " \"'library_construction': 'smart-seq2'}\\n\"\n",
      " ' ')\n",
      "Content: project_id ; Name: projectId \n",
      "\n",
      "(\"THE LEAF _file is: {'core': {'type': 'project', 'schema_url': \"\n",
      " \"'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project.json'}, \"\n",
      " \"'project_id': 'Q4_DEMO-project_PRJNA248302', 'name': 'Q4_DEMO-Single cell \"\n",
      " \"RNA-seq of primary human glioblastomas', 'contributors': [{'email': \"\n",
      " \"'mfreeberg@ebi.ac.uk', 'name': 'Mallory Freeberg'}], 'description': \"\n",
      " \"'Q4_DEMO-We report transcriptomes from 430 single glioblastoma cells \"\n",
      " 'isolated from 5 individual tumors and 102 single cells from gliomasphere '\n",
      " 'cells lines generated using SMART-seq. In addition, we report population '\n",
      " 'RNA-seq from the five tumors as well as RNA-seq from cell lines derived from '\n",
      " '3 tumors (MGH26, MGH28, MGH31) cultured under serum free (GSC) and '\n",
      " 'differentiated (DGC) conditions. This dataset highlights intratumoral '\n",
      " 'heterogeneity with regards to the expression of de novo derived '\n",
      " 'transcriptional modules and established subtype classifiers. Overall design: '\n",
      " 'Operative specimens from five glioblastoma patients (MGH26, MGH28, MGH29, '\n",
      " 'MGH30, MGH31) were acutely dissociated, depleted for CD45+ inflammatory '\n",
      " 'cells and then sorted as single cells (576 samples). Population controls for '\n",
      " 'each tumor were isolated by sorting 2000-10000 cells and processed in '\n",
      " 'parallel (5 population control samples). Single cells from two established '\n",
      " 'cell lines, GBM6 and GBM8, were also sorted as single cells (192 samples). '\n",
      " 'SMART-seq protocol was implemented to generate single cell full length '\n",
      " 'transcriptomes (modified from Shalek, et al Nature 2013) and sequenced using '\n",
      " '25 bp paired end reads. Single cell cDNA libraries for MGH30 were '\n",
      " 'resequenced using 100 bp paired end reads to allow for isoform and splice '\n",
      " 'junction reconstruction (96 samples, annotated MGH30L). Cells were also '\n",
      " 'cultured in serum free conditions to generate gliomasphere cell lines for '\n",
      " 'MGH26, MGH28, and MGH31 (GSC) which were then differentiated using 10% serum '\n",
      " '(DGC). Population RNA-seq was performed on these samples (3 GSC, 3 DGC, 6 '\n",
      " 'total). The initial dataset included 875 RNA-seq libraries (576 single '\n",
      " 'glioblastoma cells, 96 resequenced MGH30L, 192 single gliomasphere cells, 5 '\n",
      " 'tumor population controls, 6 population libraries from GSC and DGC samples). '\n",
      " 'Data was processed as described below using RSEM for quantification of gene '\n",
      " 'expression. 5,948 genes with the highest composite expression either across '\n",
      " 'all single cells combined (average log2(TPM)>4.5) or within a single tumor '\n",
      " '(average log2(TPM)>6 in at least one tumor) were included. Cells expressing '\n",
      " 'less than 2,000 of these 5,948 genes were excluded. The final processed '\n",
      " 'dataset then included 430 primary single cell glioblastoma transcriptomes, '\n",
      " '102 single cell transcriptomes from cell lines(GBM6,GBM8), 5 population '\n",
      " 'controls (1 for each tumor), and 6 population libraries from cell lines '\n",
      " 'derived from the tumors (GSC and DGC for MGH26, MGH28 and MGH31). The final '\n",
      " 'matrix (GBM_data_matrix.txt) therefore contains 5948 rows (genes) quantified '\n",
      " 'in 543 samples (columns). Please note that the samples which are not '\n",
      " 'included in the data processing are indicated in the sample description '\n",
      " \"field.\\\\n'}\\n\"\n",
      " ' ')\n",
      "the file content is: Q4_DEMO-project_PRJNA248302\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "Content: submitters ; Name: projectSubmitters \n",
      "\n",
      "(\"THE LEAF _file is: {'core': {'type': 'project', 'schema_url': \"\n",
      " \"'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project.json'}, \"\n",
      " \"'project_id': 'Q4_DEMO-project_PRJNA248302', 'name': 'Q4_DEMO-Single cell \"\n",
      " \"RNA-seq of primary human glioblastomas', 'contributors': [{'email': \"\n",
      " \"'mfreeberg@ebi.ac.uk', 'name': 'Mallory Freeberg'}], 'description': \"\n",
      " \"'Q4_DEMO-We report transcriptomes from 430 single glioblastoma cells \"\n",
      " 'isolated from 5 individual tumors and 102 single cells from gliomasphere '\n",
      " 'cells lines generated using SMART-seq. In addition, we report population '\n",
      " 'RNA-seq from the five tumors as well as RNA-seq from cell lines derived from '\n",
      " '3 tumors (MGH26, MGH28, MGH31) cultured under serum free (GSC) and '\n",
      " 'differentiated (DGC) conditions. This dataset highlights intratumoral '\n",
      " 'heterogeneity with regards to the expression of de novo derived '\n",
      " 'transcriptional modules and established subtype classifiers. Overall design: '\n",
      " 'Operative specimens from five glioblastoma patients (MGH26, MGH28, MGH29, '\n",
      " 'MGH30, MGH31) were acutely dissociated, depleted for CD45+ inflammatory '\n",
      " 'cells and then sorted as single cells (576 samples). Population controls for '\n",
      " 'each tumor were isolated by sorting 2000-10000 cells and processed in '\n",
      " 'parallel (5 population control samples). Single cells from two established '\n",
      " 'cell lines, GBM6 and GBM8, were also sorted as single cells (192 samples). '\n",
      " 'SMART-seq protocol was implemented to generate single cell full length '\n",
      " 'transcriptomes (modified from Shalek, et al Nature 2013) and sequenced using '\n",
      " '25 bp paired end reads. Single cell cDNA libraries for MGH30 were '\n",
      " 'resequenced using 100 bp paired end reads to allow for isoform and splice '\n",
      " 'junction reconstruction (96 samples, annotated MGH30L). Cells were also '\n",
      " 'cultured in serum free conditions to generate gliomasphere cell lines for '\n",
      " 'MGH26, MGH28, and MGH31 (GSC) which were then differentiated using 10% serum '\n",
      " '(DGC). Population RNA-seq was performed on these samples (3 GSC, 3 DGC, 6 '\n",
      " 'total). The initial dataset included 875 RNA-seq libraries (576 single '\n",
      " 'glioblastoma cells, 96 resequenced MGH30L, 192 single gliomasphere cells, 5 '\n",
      " 'tumor population controls, 6 population libraries from GSC and DGC samples). '\n",
      " 'Data was processed as described below using RSEM for quantification of gene '\n",
      " 'expression. 5,948 genes with the highest composite expression either across '\n",
      " 'all single cells combined (average log2(TPM)>4.5) or within a single tumor '\n",
      " '(average log2(TPM)>6 in at least one tumor) were included. Cells expressing '\n",
      " 'less than 2,000 of these 5,948 genes were excluded. The final processed '\n",
      " 'dataset then included 430 primary single cell glioblastoma transcriptomes, '\n",
      " '102 single cell transcriptomes from cell lines(GBM6,GBM8), 5 population '\n",
      " 'controls (1 for each tumor), and 6 population libraries from cell lines '\n",
      " 'derived from the tumors (GSC and DGC for MGH26, MGH28 and MGH31). The final '\n",
      " 'matrix (GBM_data_matrix.txt) therefore contains 5948 rows (genes) quantified '\n",
      " 'in 543 samples (columns). Please note that the samples which are not '\n",
      " 'included in the data processing are indicated in the sample description '\n",
      " \"field.\\\\n'}\\n\"\n",
      " ' ')\n",
      "Content: email ; Name: projectContributosEmail \n",
      "\n",
      "(\"THE LEAF _file is: [{'email': 'mfreeberg@ebi.ac.uk', 'name': 'Mallory \"\n",
      " \"Freeberg'}]\\n\"\n",
      " ' ')\n",
      "PRINTING LEAF el\n",
      "{'email': 'mfreeberg@ebi.ac.uk', 'name': 'Mallory Freeberg'}\n",
      "Content: email ; Name: projectContributosEmail \n",
      "\n",
      "(\"THE LEAF _file is: {'email': 'mfreeberg@ebi.ac.uk', 'name': 'Mallory \"\n",
      " \"Freeberg'}\\n\"\n",
      " ' ')\n",
      "the file content is: mfreeberg@ebi.ac.uk\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectContributosEmail': ['mfreeberg@ebi.ac.uk'],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "PRINTING _file:\n",
      "[{'content': {'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                       'schema_version': '4.6.1',\n",
      "                       'type': 'sample'},\n",
      "              'derived_from': 'Q4_DEMO-donor_MGH30',\n",
      "              'genus_species': {'ontology': 'NCBITaxon:9606',\n",
      "                                'text': 'Homo sapiens'},\n",
      "              'name': 'Q4_DEMO-Single cell mRNA-seq_MGH30_A01',\n",
      "              'ncbi_taxon_id': 9606,\n",
      "              'sample_id': 'Q4_DEMO-sample_SAMN02797092',\n",
      "              'specimen_from_organism': {'body_part': {'text': 'glioblastoma'},\n",
      "                                         'organ': {'ontology': 'UBERON:0000955',\n",
      "                                                   'text': 'brain'}}},\n",
      "  'derivation_protocols': [{'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/protocol.json',\n",
      "                                     'schema_version': '4.6.1',\n",
      "                                     'type': 'protocol'},\n",
      "                            'pdf': 'Q4_DEMO-protocol.pdf',\n",
      "                            'protocol_id': 'Q4_DEMO-protocol',\n",
      "                            'type': {'text': 'single cell sequencing'}}],\n",
      "  'derived_from': '176ff29c-e492-495b-a537-d21940a53799',\n",
      "  'hca_ingest': {'accession': '',\n",
      "                 'document_id': '2e377554-32bb-4ded-abc9-a8ada91798de',\n",
      "                 'lastModifiedUser': 'anonymousUser',\n",
      "                 'submissionDate': '2018-01-03T17:16:34.953Z',\n",
      "                 'updateDate': '2018-01-03T17:17:17.116Z',\n",
      "                 'user': 'anonymousUser'}},\n",
      " {'content': {'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                       'schema_version': '4.6.1',\n",
      "                       'type': 'sample'},\n",
      "              'donor': {'is_living': True},\n",
      "              'genus_species': {'ontology': 'NCBITaxon:9606',\n",
      "                                'text': 'Homo sapiens'},\n",
      "              'name': 'Q4 DEMO donor MGH30',\n",
      "              'ncbi_taxon_id': 9606,\n",
      "              'sample_id': 'Q4_DEMO-donor_MGH30'},\n",
      "  'derivation_protocols': [],\n",
      "  'hca_ingest': {'accession': '',\n",
      "                 'document_id': '176ff29c-e492-495b-a537-d21940a53799',\n",
      "                 'lastModifiedUser': 'anonymousUser',\n",
      "                 'submissionDate': '2018-01-03T17:16:34.938Z',\n",
      "                 'updateDate': '2018-01-03T17:17:15.915Z',\n",
      "                 'user': 'anonymousUser'}}]\n",
      "PRINTING el\n",
      "{'content': {'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                      'schema_version': '4.6.1',\n",
      "                      'type': 'sample'},\n",
      "             'derived_from': 'Q4_DEMO-donor_MGH30',\n",
      "             'genus_species': {'ontology': 'NCBITaxon:9606',\n",
      "                               'text': 'Homo sapiens'},\n",
      "             'name': 'Q4_DEMO-Single cell mRNA-seq_MGH30_A01',\n",
      "             'ncbi_taxon_id': 9606,\n",
      "             'sample_id': 'Q4_DEMO-sample_SAMN02797092',\n",
      "             'specimen_from_organism': {'body_part': {'text': 'glioblastoma'},\n",
      "                                        'organ': {'ontology': 'UBERON:0000955',\n",
      "                                                  'text': 'brain'}}},\n",
      " 'derivation_protocols': [{'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/protocol.json',\n",
      "                                    'schema_version': '4.6.1',\n",
      "                                    'type': 'protocol'},\n",
      "                           'pdf': 'Q4_DEMO-protocol.pdf',\n",
      "                           'protocol_id': 'Q4_DEMO-protocol',\n",
      "                           'type': {'text': 'single cell sequencing'}}],\n",
      " 'derived_from': '176ff29c-e492-495b-a537-d21940a53799',\n",
      " 'hca_ingest': {'accession': '',\n",
      "                'document_id': '2e377554-32bb-4ded-abc9-a8ada91798de',\n",
      "                'lastModifiedUser': 'anonymousUser',\n",
      "                'submissionDate': '2018-01-03T17:16:34.953Z',\n",
      "                'updateDate': '2018-01-03T17:17:17.116Z',\n",
      "                'user': 'anonymousUser'}}\n",
      "Content: text ; Name: sampleBodyPart \n",
      "\n",
      "\"THE LEAF _file is: {'text': 'glioblastoma'}\\n \"\n",
      "the file content is: glioblastoma\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectContributosEmail': ['mfreeberg@ebi.ac.uk'],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'sampleBodyPart': ['glioblastoma'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "Content: text ; Name: sampleSpecies \n",
      "\n",
      "\"THE LEAF _file is: {'text': 'Homo sapiens', 'ontology': 'NCBITaxon:9606'}\\n \"\n",
      "the file content is: Homo sapiens\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectContributosEmail': ['mfreeberg@ebi.ac.uk'],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'sampleBodyPart': ['glioblastoma'],\n",
      "             'sampleSpecies': ['Homo sapiens'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "PRINTING el\n",
      "{'content': {'core': {'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                      'schema_version': '4.6.1',\n",
      "                      'type': 'sample'},\n",
      "             'donor': {'is_living': True},\n",
      "             'genus_species': {'ontology': 'NCBITaxon:9606',\n",
      "                               'text': 'Homo sapiens'},\n",
      "             'name': 'Q4 DEMO donor MGH30',\n",
      "             'ncbi_taxon_id': 9606,\n",
      "             'sample_id': 'Q4_DEMO-donor_MGH30'},\n",
      " 'derivation_protocols': [],\n",
      " 'hca_ingest': {'accession': '',\n",
      "                'document_id': '176ff29c-e492-495b-a537-d21940a53799',\n",
      "                'lastModifiedUser': 'anonymousUser',\n",
      "                'submissionDate': '2018-01-03T17:16:34.938Z',\n",
      "                'updateDate': '2018-01-03T17:17:15.915Z',\n",
      "                'user': 'anonymousUser'}}\n",
      "Content: text ; Name: sampleSpecies \n",
      "\n",
      "\"THE LEAF _file is: {'text': 'Homo sapiens', 'ontology': 'NCBITaxon:9606'}\\n \"\n",
      "the file content is: Homo sapiens\n",
      "\n",
      "JUST APPENDED ENTRY, dictionary right now:\n",
      "\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectContributosEmail': ['mfreeberg@ebi.ac.uk'],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'sampleBodyPart': ['glioblastoma'],\n",
      "             'sampleSpecies': ['Homo sapiens', 'Homo sapiens'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "#######TESTING NEW get_item2 METHOD##########\n",
      "defaultdict(<class 'list'>,\n",
      "            {'assayId': ['assay_1'],\n",
      "             'pairedEnds': [True],\n",
      "             'projectContributosEmail': ['mfreeberg@ebi.ac.uk'],\n",
      "             'projectId': ['Q4_DEMO-project_PRJNA248302'],\n",
      "             'rnaLibraryConstruction': ['smart-seq2'],\n",
      "             'sampleBodyPart': ['glioblastoma'],\n",
      "             'sampleSpecies': ['Homo sapiens', 'Homo sapiens'],\n",
      "             'scCellHandling': ['FACS'],\n",
      "             'seqInstrumentPlatform': ['Illumina']})\n",
      "INDEXING DONE\n"
     ]
    }
   ],
   "source": [
    "# Define helper method to open files\n",
    "def open_and_return_json(file_path):\n",
    "    \"\"\"\n",
    "    Opens and returns the contents of the json file given in file_path\n",
    "    :param file_path: Path of a json file to be opened\n",
    "    :return: Returns an obj with the contents of the json file\n",
    "    http://localhost:8888/notebooks/Indexing_Code_Tour.ipynb#\"\"\"\n",
    "    with open(file_path, 'r') as file_:\n",
    "        loaded_file = json.load(file_)\n",
    "    return loaded_file\n",
    "\n",
    "# Get the index's settings\n",
    "index_settings = open_and_return_json('chalicelib/settings.json')\n",
    "# Get the index overall config\n",
    "index_mapping_config = open_and_return_json('chalicelib/config.json')\n",
    "\n",
    "file_indexer = FileIndexer(metadata_files,\n",
    "                           data_files,\n",
    "                           es,\n",
    "                           \"file_index_v4\",\n",
    "                           \"doc\",\n",
    "                           index_settings=index_settings,\n",
    "                           index_mapping_config=index_mapping_config)\n",
    "\n",
    "assay_indexer = AssayOrientedIndexer(metadata_files,\n",
    "                                     data_files,\n",
    "                                     es,\n",
    "                                     \"assay_index_v4\",\n",
    "                                     \"doc\",\n",
    "                                     index_settings=index_settings,\n",
    "                                     index_mapping_config=index_mapping_config)\n",
    "\n",
    "file_indexer.index(bundle_uuid, bundle_version)\n",
    "#assay_indexer.index(bundle_uuid, bundle_version)\n",
    "print(\"INDEXING DONE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
