{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DSS-AZUL-INDEXING\n",
    "\n",
    "### This jupyter notebook is to do a couple testing indexing operations using the DataExtractor class and the FileIndexer class."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below here we import our modules and set up:\n",
    "* ElasticSearch Client\n",
    "* Dummy payload of event\n",
    "* Parse the bundle_uuid and the bundle_version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from elasticsearch import Elasticsearch\n",
    "from indexer.utils import DataExtractor\n",
    "from indexer.indexer import FileIndexer, AssayOrientedIndexer\n",
    "import json\n",
    "from pprint import pprint\n",
    "\n",
    "# Create an ElasticSearch client\n",
    "es = Elasticsearch([{'host': 'localhost', 'port': 9200}])\n",
    "# Sample event payload\n",
    "payload = { \"query\": { \"query\": { \"match_all\":{}} }, \"subscription_id\": \"ba50df7b-5a97-4e87-b9ce-c0935a817f0b\", \"transaction_id\": \"ff6b7fa3-dc79-4a79-a313-296801de76b9\", \"match\": { \"bundle_version\": \"2017-12-21T164607.645085Z\", \"bundle_uuid\": \"a4aed7f3-7fbf-45ee-b2aa-4402126414ab\" } }\n",
    "bundle_uuid = payload['match']['bundle_uuid']\n",
    "bundle_version = payload['match']['bundle_version']\n",
    "# Create "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will create an instance of the DataExtractor and use it to get the contents from the bundle referenced by the variable `payload`. We will be pulling from the AWS replica."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "#####################################################\n",
      "#                    PRINTING METADATA               #\n",
      "#####################################################\n",
      "{   'assay.json': {   'content': {   'assay_id': 'assay_1',\n",
      "                                     'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/assay.json',\n",
      "                                                 'schema_version': '4.6.1',\n",
      "                                                 'type': 'assay'},\n",
      "                                     'rna': {   'end_bias': 'five_prime_end',\n",
      "                                                'library_construction': 'smart-seq2',\n",
      "                                                'strand': 'both'},\n",
      "                                     'seq': {   'instrument_model': 'HiSeq '\n",
      "                                                                    '2500',\n",
      "                                                'instrument_platform': 'Illumina',\n",
      "                                                'lanes': [   {   'number': 1,\n",
      "                                                                 'r1': 'R1.fastq.gz',\n",
      "                                                                 'r2': 'R2.fastq.gz'}],\n",
      "                                                'molecule': 'polyA RNA',\n",
      "                                                'paired_ends': True},\n",
      "                                     'single_cell': {'cell_handling': 'FACS'}},\n",
      "                      'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/assay_bundle.json',\n",
      "                                  'schema_version': '4.6.1',\n",
      "                                  'type': 'assay_bundle'},\n",
      "                      'has_input': '243944a6-546b-4f84-ba61-64a041f7d29e',\n",
      "                      'has_output': [   '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d',\n",
      "                                        'cedf25a8-ce28-4025-a686-459b220fc825'],\n",
      "                      'hca_ingest': {   'accession': '',\n",
      "                                        'document_id': 'aaf29962-e84b-4b14-a9ef-8de9e54122f1',\n",
      "                                        'lastModifiedUser': 'anonymousUser',\n",
      "                                        'submissionDate': '2017-12-21T16:44:39.794Z',\n",
      "                                        'updateDate': '2017-12-21T16:45:25.501Z',\n",
      "                                        'user': 'anonymousUser'}},\n",
      "    'project.json': {   'content': {   'contributors': [   {   'email': 'dummy@email.com',\n",
      "                                                               'name': 'Q3_DEMO-MintTeam'}],\n",
      "                                       'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project.json',\n",
      "                                                   'schema_version': '4.6.1',\n",
      "                                                   'type': 'project'},\n",
      "                                       'description': 'Q3_DEMO-We report '\n",
      "                                                      'transcriptomes from 430 '\n",
      "                                                      'single glioblastoma '\n",
      "                                                      'cells isolated from 5 '\n",
      "                                                      'individual tumors and '\n",
      "                                                      '102 single cells from '\n",
      "                                                      'gliomasphere cells '\n",
      "                                                      'lines generated using '\n",
      "                                                      'SMART-seq. In addition, '\n",
      "                                                      'we report population '\n",
      "                                                      'RNA-seq from the five '\n",
      "                                                      'tumors as well as '\n",
      "                                                      'RNA-seq from cell lines '\n",
      "                                                      'derived from 3 tumors '\n",
      "                                                      '(MGH26, MGH28, MGH31) '\n",
      "                                                      'cultured under serum '\n",
      "                                                      'free (GSC) and '\n",
      "                                                      'differentiated (DGC) '\n",
      "                                                      'conditions. This '\n",
      "                                                      'dataset highlights '\n",
      "                                                      'intratumoral '\n",
      "                                                      'heterogeneity with '\n",
      "                                                      'regards to the '\n",
      "                                                      'expression of de novo '\n",
      "                                                      'derived transcriptional '\n",
      "                                                      'modules and established '\n",
      "                                                      'subtype classifiers. '\n",
      "                                                      'Overall design: '\n",
      "                                                      'Operative specimens '\n",
      "                                                      'from five glioblastoma '\n",
      "                                                      'patients (MGH26, MGH28, '\n",
      "                                                      'MGH29, MGH30, MGH31) '\n",
      "                                                      'were acutely '\n",
      "                                                      'dissociated, depleted '\n",
      "                                                      'for CD45+ inflammatory '\n",
      "                                                      'cells and then sorted '\n",
      "                                                      'as single cells (576 '\n",
      "                                                      'samples). Population '\n",
      "                                                      'controls for each tumor '\n",
      "                                                      'were isolated by '\n",
      "                                                      'sorting 2000-10000 '\n",
      "                                                      'cells and processed in '\n",
      "                                                      'parallel (5 population '\n",
      "                                                      'control samples). '\n",
      "                                                      'Single cells from two '\n",
      "                                                      'established cell lines, '\n",
      "                                                      'GBM6 and GBM8, were '\n",
      "                                                      'also sorted as single '\n",
      "                                                      'cells (192 samples). '\n",
      "                                                      'SMART-seq protocol was '\n",
      "                                                      'implemented to generate '\n",
      "                                                      'single cell full length '\n",
      "                                                      'transcriptomes '\n",
      "                                                      '(modified from Shalek, '\n",
      "                                                      'et al Nature 2013) and '\n",
      "                                                      'sequenced using 25 bp '\n",
      "                                                      'paired end reads. '\n",
      "                                                      'Single cell cDNA '\n",
      "                                                      'libraries for MGH30 '\n",
      "                                                      'were resequenced using '\n",
      "                                                      '100 bp paired end reads '\n",
      "                                                      'to allow for isoform '\n",
      "                                                      'and splice junction '\n",
      "                                                      'reconstruction (96 '\n",
      "                                                      'samples, annotated '\n",
      "                                                      'MGH30L). Cells were '\n",
      "                                                      'also cultured in serum '\n",
      "                                                      'free conditions to '\n",
      "                                                      'generate gliomasphere '\n",
      "                                                      'cell lines for MGH26, '\n",
      "                                                      'MGH28, and MGH31 (GSC) '\n",
      "                                                      'which were then '\n",
      "                                                      'differentiated using '\n",
      "                                                      '10% serum (DGC). '\n",
      "                                                      'Population RNA-seq was '\n",
      "                                                      'performed on these '\n",
      "                                                      'samples (3 GSC, 3 DGC, '\n",
      "                                                      '6 total). The initial '\n",
      "                                                      'dataset included 875 '\n",
      "                                                      'RNA-seq libraries (576 '\n",
      "                                                      'single glioblastoma '\n",
      "                                                      'cells, 96 resequenced '\n",
      "                                                      'MGH30L, 192 single '\n",
      "                                                      'gliomasphere cells, 5 '\n",
      "                                                      'tumor population '\n",
      "                                                      'controls, 6 population '\n",
      "                                                      'libraries from GSC and '\n",
      "                                                      'DGC samples). Data was '\n",
      "                                                      'processed as described '\n",
      "                                                      'below using RSEM for '\n",
      "                                                      'quantification of gene '\n",
      "                                                      'expression. 5,948 genes '\n",
      "                                                      'with the highest '\n",
      "                                                      'composite expression '\n",
      "                                                      'either across all '\n",
      "                                                      'single cells combined '\n",
      "                                                      '(average log2(TPM)>4.5) '\n",
      "                                                      'or within a single '\n",
      "                                                      'tumor (average '\n",
      "                                                      'log2(TPM)>6 in at least '\n",
      "                                                      'one tumor) were '\n",
      "                                                      'included. Cells '\n",
      "                                                      'expressing less than '\n",
      "                                                      '2,000 of these 5,948 '\n",
      "                                                      'genes were excluded. '\n",
      "                                                      'The final processed '\n",
      "                                                      'dataset then included '\n",
      "                                                      '430 primary single cell '\n",
      "                                                      'glioblastoma '\n",
      "                                                      'transcriptomes, 102 '\n",
      "                                                      'single cell '\n",
      "                                                      'transcriptomes from '\n",
      "                                                      'cell lines(GBM6,GBM8), '\n",
      "                                                      '5 population controls '\n",
      "                                                      '(1 for each tumor), and '\n",
      "                                                      '6 population libraries '\n",
      "                                                      'from cell lines derived '\n",
      "                                                      'from the tumors (GSC '\n",
      "                                                      'and DGC for MGH26, '\n",
      "                                                      'MGH28 and MGH31). The '\n",
      "                                                      'final matrix '\n",
      "                                                      '(GBM_data_matrix.txt) '\n",
      "                                                      'therefore contains 5948 '\n",
      "                                                      'rows (genes) quantified '\n",
      "                                                      'in 543 samples '\n",
      "                                                      '(columns). Please note '\n",
      "                                                      'that the samples which '\n",
      "                                                      'are not included in the '\n",
      "                                                      'data processing are '\n",
      "                                                      'indicated in the sample '\n",
      "                                                      'description field.',\n",
      "                                       'name': 'Q3_DEMO-Single cell RNA-seq of '\n",
      "                                               'primary human glioblastomas',\n",
      "                                       'project_id': 'Q3_DEMO-project_PRJNA248302',\n",
      "                                       'publications': [],\n",
      "                                       'submitters': [   {   'email': 'dummy@email.com',\n",
      "                                                             'name': 'Q3_DEMO-MintTeam'}]},\n",
      "                        'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project_bundle.json',\n",
      "                                    'schema_version': '4.6.1',\n",
      "                                    'type': 'project_bundle'},\n",
      "                        'hca_ingest': {   'accession': '',\n",
      "                                          'document_id': 'abc0cd74-45ce-4199-bc25-734cf09fec13',\n",
      "                                          'lastModifiedUser': 'anonymousUser',\n",
      "                                          'submissionDate': '2017-12-21T16:44:39.684Z',\n",
      "                                          'updateDate': '2017-12-21T16:45:21.203Z',\n",
      "                                          'user': 'anonymousUser'}},\n",
      "    'sample.json': {   'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample_bundle.json',\n",
      "                                   'schema_version': '4.6.1',\n",
      "                                   'type': 'sample_bundle'},\n",
      "                       'samples': [   {   'content': {   'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                                                                     'schema_version': '4.6.1',\n",
      "                                                                     'type': 'sample'},\n",
      "                                                         'derived_from': 'Q3_DEMO-donor_MGH30',\n",
      "                                                         'name': 'Q3_DEMO-Single '\n",
      "                                                                 'cell '\n",
      "                                                                 'mRNA-seq_MGH30_A01',\n",
      "                                                         'ncbi_taxon_id': 9606,\n",
      "                                                         'sample_id': 'Q3_DEMO-sample_SAMN02797092',\n",
      "                                                         'specimen_from_organism': {   'body_part': {   'text': 'glioblastoma'},\n",
      "                                                                                       'organ': {   'text': 'brain'}}},\n",
      "                                          'derivation_protocols': [   {   'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/protocol.json',\n",
      "                                                                                      'schema_version': '4.6.1',\n",
      "                                                                                      'type': 'protocol'},\n",
      "                                                                          'pdf': 'Q3_DEMO-protocol.pdf',\n",
      "                                                                          'protocol_id': 'Q3_DEMO-protocol',\n",
      "                                                                          'type': {   'text': 'single '\n",
      "                                                                                              'cell '\n",
      "                                                                                              'sequencing'}}],\n",
      "                                          'derived_from': '22f06c77-539e-460b-a023-80da55a0b8ff',\n",
      "                                          'hca_ingest': {   'accession': '',\n",
      "                                                            'document_id': '243944a6-546b-4f84-ba61-64a041f7d29e',\n",
      "                                                            'lastModifiedUser': 'anonymousUser',\n",
      "                                                            'submissionDate': '2017-12-21T16:44:39.726Z',\n",
      "                                                            'updateDate': '2017-12-21T16:45:20.475Z',\n",
      "                                                            'user': 'anonymousUser'}},\n",
      "                                      {   'content': {   'core': {   'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json',\n",
      "                                                                     'schema_version': '4.6.1',\n",
      "                                                                     'type': 'sample'},\n",
      "                                                         'donor': {   'is_living': True},\n",
      "                                                         'genus_species': {   'ontology': 'NCBITaxon:9606',\n",
      "                                                                              'text': 'Homo '\n",
      "                                                                                      'sapiens'},\n",
      "                                                         'name': 'Q3 DEMO '\n",
      "                                                                 'donor MGH30',\n",
      "                                                         'ncbi_taxon_id': 9606,\n",
      "                                                         'sample_id': 'Q3_DEMO-donor_MGH30'},\n",
      "                                          'derivation_protocols': [],\n",
      "                                          'hca_ingest': {   'accession': '',\n",
      "                                                            'document_id': '22f06c77-539e-460b-a023-80da55a0b8ff',\n",
      "                                                            'lastModifiedUser': 'anonymousUser',\n",
      "                                                            'submissionDate': '2017-12-21T16:44:39.707Z',\n",
      "                                                            'updateDate': '2017-12-21T16:45:20.601Z',\n",
      "                                                            'user': 'anonymousUser'}}]}}\n",
      "\n",
      "#####################################################\n",
      "#                  PRINTING DATA FILES               #\n",
      "#####################################################\n",
      "{   'R1.fastq.gz': {   'content-type': 'application/gzip; dcp-type=data',\n",
      "                       'crc32c': '4ef74578',\n",
      "                       'indexed': False,\n",
      "                       'name': 'R1.fastq.gz',\n",
      "                       's3_etag': 'c7bbee4c46bbf29432862e05830c8f39',\n",
      "                       'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3',\n",
      "                       'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29',\n",
      "                       'size': 125191,\n",
      "                       'uuid': '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d',\n",
      "                       'version': '2017-12-21T164604.038618Z'},\n",
      "    'R2.fastq.gz': {   'content-type': 'application/gzip; dcp-type=data',\n",
      "                       'crc32c': '69987b3e',\n",
      "                       'indexed': False,\n",
      "                       'name': 'R2.fastq.gz',\n",
      "                       's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae',\n",
      "                       'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21',\n",
      "                       'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba',\n",
      "                       'size': 130024,\n",
      "                       'uuid': 'cedf25a8-ce28-4025-a686-459b220fc825',\n",
      "                       'version': '2017-12-21T164605.003656Z'}}\n"
     ]
    }
   ],
   "source": [
    "# Create DataExtractor instance pointing to HCA Staging\n",
    "extractor = DataExtractor(\"https://dss.staging.data.humancellatlas.org/v1\")\n",
    "# Use dummy payload and get the metadata_files and the data_files\n",
    "metadata_files, data_files = extractor.extract_bundle(payload, \"aws\")\n",
    "# Print each dictionary\n",
    "print(\"\\n#####################################################\")\n",
    "print(\"#                    PRINTING METADATA               #\")\n",
    "print(\"#####################################################\")\n",
    "pprint(metadata_files, indent=4)\n",
    "print(\"\\n#####################################################\")\n",
    "print(\"#                  PRINTING DATA FILES               #\")\n",
    "print(\"#####################################################\")\n",
    "pprint(data_files, indent=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we pass this on to the FileIndexer class to create a File Oriented index entry on ElasticSearch running on `localhost:9200`. But first, we get the index settings and get the configuration files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#########PRINTING ITEM IN __merge_lists###############:\n",
      "{'content': {'core': {'type': 'sample', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json', 'schema_version': '4.6.1'}, 'name': 'Q3_DEMO-Single cell mRNA-seq_MGH30_A01', 'specimen_from_organism': {'body_part': {'text': 'glioblastoma'}, 'organ': {'text': 'brain'}}, 'ncbi_taxon_id': 9606, 'derived_from': 'Q3_DEMO-donor_MGH30', 'sample_id': 'Q3_DEMO-sample_SAMN02797092'}, 'derivation_protocols': [{'pdf': 'Q3_DEMO-protocol.pdf', 'protocol_id': 'Q3_DEMO-protocol', 'type': {'text': 'single cell sequencing'}, 'core': {'type': 'protocol', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/protocol.json', 'schema_version': '4.6.1'}}], 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.726Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:20.475Z', 'document_id': '243944a6-546b-4f84-ba61-64a041f7d29e', 'user': 'anonymousUser'}, 'derived_from': '22f06c77-539e-460b-a023-80da55a0b8ff'}\n",
      "\n",
      "<class 'dict'>\n",
      "####GET_VALUE d:####\n",
      "{'content': {'core': {'type': 'sample', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json', 'schema_version': '4.6.1'}, 'name': 'Q3_DEMO-Single cell mRNA-seq_MGH30_A01', 'specimen_from_organism': {'body_part': {'text': 'glioblastoma'}, 'organ': {'text': 'brain'}}, 'ncbi_taxon_id': 9606, 'derived_from': 'Q3_DEMO-donor_MGH30', 'sample_id': 'Q3_DEMO-sample_SAMN02797092'}, 'derivation_protocols': [{'pdf': 'Q3_DEMO-protocol.pdf', 'protocol_id': 'Q3_DEMO-protocol', 'type': {'text': 'single cell sequencing'}, 'core': {'type': 'protocol', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/protocol.json', 'schema_version': '4.6.1'}}], 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.726Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:20.475Z', 'document_id': '243944a6-546b-4f84-ba61-64a041f7d29e', 'user': 'anonymousUser'}, 'derived_from': '22f06c77-539e-460b-a023-80da55a0b8ff'}\n",
      "\n",
      "#########PRINTING ITEM IN __merge_lists###############:\n",
      "{'content': {'core': {'type': 'sample', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json', 'schema_version': '4.6.1'}, 'name': 'Q3 DEMO donor MGH30', 'genus_species': {'text': 'Homo sapiens', 'ontology': 'NCBITaxon:9606'}, 'ncbi_taxon_id': 9606, 'sample_id': 'Q3_DEMO-donor_MGH30', 'donor': {'is_living': True}}, 'derivation_protocols': [], 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.707Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:20.601Z', 'document_id': '22f06c77-539e-460b-a023-80da55a0b8ff', 'user': 'anonymousUser'}}\n",
      "\n",
      "<class 'dict'>\n",
      "####GET_VALUE d:####\n",
      "{'content': {'core': {'type': 'sample', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json', 'schema_version': '4.6.1'}, 'name': 'Q3 DEMO donor MGH30', 'genus_species': {'text': 'Homo sapiens', 'ontology': 'NCBITaxon:9606'}, 'ncbi_taxon_id': 9606, 'sample_id': 'Q3_DEMO-donor_MGH30', 'donor': {'is_living': True}}, 'derivation_protocols': [], 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.707Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:20.601Z', 'document_id': '22f06c77-539e-460b-a023-80da55a0b8ff', 'user': 'anonymousUser'}}\n",
      "\n",
      "#########PRINTING ITEM IN __merge_lists###############:\n",
      "{'content': {'core': {'type': 'sample', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json', 'schema_version': '4.6.1'}, 'name': 'Q3_DEMO-Single cell mRNA-seq_MGH30_A01', 'specimen_from_organism': {'body_part': {'text': 'glioblastoma'}, 'organ': {'text': 'brain'}}, 'ncbi_taxon_id': 9606, 'derived_from': 'Q3_DEMO-donor_MGH30', 'sample_id': 'Q3_DEMO-sample_SAMN02797092'}, 'derivation_protocols': [{'pdf': 'Q3_DEMO-protocol.pdf', 'protocol_id': 'Q3_DEMO-protocol', 'type': {'text': 'single cell sequencing'}, 'core': {'type': 'protocol', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/protocol.json', 'schema_version': '4.6.1'}}], 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.726Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:20.475Z', 'document_id': '243944a6-546b-4f84-ba61-64a041f7d29e', 'user': 'anonymousUser'}, 'derived_from': '22f06c77-539e-460b-a023-80da55a0b8ff'}\n",
      "\n",
      "<class 'dict'>\n",
      "####GET_VALUE d:####\n",
      "{'content': {'core': {'type': 'sample', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json', 'schema_version': '4.6.1'}, 'name': 'Q3_DEMO-Single cell mRNA-seq_MGH30_A01', 'specimen_from_organism': {'body_part': {'text': 'glioblastoma'}, 'organ': {'text': 'brain'}}, 'ncbi_taxon_id': 9606, 'derived_from': 'Q3_DEMO-donor_MGH30', 'sample_id': 'Q3_DEMO-sample_SAMN02797092'}, 'derivation_protocols': [{'pdf': 'Q3_DEMO-protocol.pdf', 'protocol_id': 'Q3_DEMO-protocol', 'type': {'text': 'single cell sequencing'}, 'core': {'type': 'protocol', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/protocol.json', 'schema_version': '4.6.1'}}], 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.726Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:20.475Z', 'document_id': '243944a6-546b-4f84-ba61-64a041f7d29e', 'user': 'anonymousUser'}, 'derived_from': '22f06c77-539e-460b-a023-80da55a0b8ff'}\n",
      "\n",
      "#########PRINTING ITEM IN __merge_lists###############:\n",
      "{'content': {'core': {'type': 'sample', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json', 'schema_version': '4.6.1'}, 'name': 'Q3 DEMO donor MGH30', 'genus_species': {'text': 'Homo sapiens', 'ontology': 'NCBITaxon:9606'}, 'ncbi_taxon_id': 9606, 'sample_id': 'Q3_DEMO-donor_MGH30', 'donor': {'is_living': True}}, 'derivation_protocols': [], 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.707Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:20.601Z', 'document_id': '22f06c77-539e-460b-a023-80da55a0b8ff', 'user': 'anonymousUser'}}\n",
      "\n",
      "<class 'dict'>\n",
      "####GET_VALUE d:####\n",
      "{'content': {'core': {'type': 'sample', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/sample.json', 'schema_version': '4.6.1'}, 'name': 'Q3 DEMO donor MGH30', 'genus_species': {'text': 'Homo sapiens', 'ontology': 'NCBITaxon:9606'}, 'ncbi_taxon_id': 9606, 'sample_id': 'Q3_DEMO-donor_MGH30', 'donor': {'is_living': True}}, 'derivation_protocols': [], 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.707Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:20.601Z', 'document_id': '22f06c77-539e-460b-a023-80da55a0b8ff', 'user': 'anonymousUser'}}\n",
      "\n",
      "#########PRINTING ITEM IN __merge_lists###############:\n",
      "{'content': {'analysis_id': 'assay_1'}, 'analysis_type': 'upload', 'data_bundles': [{'bundle_uuid': 'a4aed7f3-7fbf-45ee-b2aa-4402126414ab', 'files': [{'content-type': 'application/gzip; dcp-type=data', 'crc32c': '4ef74578', 'indexed': False, 'name': 'R1.fastq.gz', 's3_etag': 'c7bbee4c46bbf29432862e05830c8f39', 'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3', 'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29', 'size': 125191, 'uuid': '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d', 'version': '2017-12-21T164604.038618Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}, {'content-type': 'application/gzip; dcp-type=data', 'crc32c': '69987b3e', 'indexed': False, 'name': 'R2.fastq.gz', 's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae', 'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21', 'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba', 'size': 130024, 'uuid': 'cedf25a8-ce28-4025-a686-459b220fc825', 'version': '2017-12-21T164605.003656Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}]}]}\n",
      "\n",
      "<class 'dict'>\n",
      "####GET_VALUE d:####\n",
      "{'content': {'analysis_id': 'assay_1'}, 'analysis_type': 'upload', 'data_bundles': [{'bundle_uuid': 'a4aed7f3-7fbf-45ee-b2aa-4402126414ab', 'files': [{'content-type': 'application/gzip; dcp-type=data', 'crc32c': '4ef74578', 'indexed': False, 'name': 'R1.fastq.gz', 's3_etag': 'c7bbee4c46bbf29432862e05830c8f39', 'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3', 'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29', 'size': 125191, 'uuid': '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d', 'version': '2017-12-21T164604.038618Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}, {'content-type': 'application/gzip; dcp-type=data', 'crc32c': '69987b3e', 'indexed': False, 'name': 'R2.fastq.gz', 's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae', 'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21', 'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba', 'size': 130024, 'uuid': 'cedf25a8-ce28-4025-a686-459b220fc825', 'version': '2017-12-21T164605.003656Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}]}]}\n",
      "\n",
      "#########PRINTING ITEM IN __merge_lists###############:\n",
      "{'content': {'analysis_id': 'assay_1'}, 'analysis_type': 'upload', 'data_bundles': [{'bundle_uuid': 'a4aed7f3-7fbf-45ee-b2aa-4402126414ab', 'files': [{'content-type': 'application/gzip; dcp-type=data', 'crc32c': '4ef74578', 'indexed': False, 'name': 'R1.fastq.gz', 's3_etag': 'c7bbee4c46bbf29432862e05830c8f39', 'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3', 'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29', 'size': 125191, 'uuid': '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d', 'version': '2017-12-21T164604.038618Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}, {'content-type': 'application/gzip; dcp-type=data', 'crc32c': '69987b3e', 'indexed': False, 'name': 'R2.fastq.gz', 's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae', 'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21', 'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba', 'size': 130024, 'uuid': 'cedf25a8-ce28-4025-a686-459b220fc825', 'version': '2017-12-21T164605.003656Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}]}]}\n",
      "\n",
      "<class 'dict'>\n",
      "####GET_VALUE d:####\n",
      "{'content': {'analysis_id': 'assay_1'}, 'analysis_type': 'upload', 'data_bundles': [{'bundle_uuid': 'a4aed7f3-7fbf-45ee-b2aa-4402126414ab', 'files': [{'content-type': 'application/gzip; dcp-type=data', 'crc32c': '4ef74578', 'indexed': False, 'name': 'R1.fastq.gz', 's3_etag': 'c7bbee4c46bbf29432862e05830c8f39', 'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3', 'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29', 'size': 125191, 'uuid': '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d', 'version': '2017-12-21T164604.038618Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}, {'content-type': 'application/gzip; dcp-type=data', 'crc32c': '69987b3e', 'indexed': False, 'name': 'R2.fastq.gz', 's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae', 'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21', 'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba', 'size': 130024, 'uuid': 'cedf25a8-ce28-4025-a686-459b220fc825', 'version': '2017-12-21T164605.003656Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}]}]}\n",
      "\n",
      "####GET_VALUE d:####\n",
      "{'content': {'analysis_id': 'assay_1'}, 'analysis_type': 'upload', 'data_bundles': [{'bundle_uuid': 'a4aed7f3-7fbf-45ee-b2aa-4402126414ab', 'files': [{'content-type': 'application/gzip; dcp-type=data', 'crc32c': '4ef74578', 'indexed': False, 'name': 'R1.fastq.gz', 's3_etag': 'c7bbee4c46bbf29432862e05830c8f39', 'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3', 'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29', 'size': 125191, 'uuid': '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d', 'version': '2017-12-21T164604.038618Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}, {'content-type': 'application/gzip; dcp-type=data', 'crc32c': '69987b3e', 'indexed': False, 'name': 'R2.fastq.gz', 's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae', 'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21', 'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba', 'size': 130024, 'uuid': 'cedf25a8-ce28-4025-a686-459b220fc825', 'version': '2017-12-21T164605.003656Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}]}]}\n",
      "\n",
      "####GET_VALUE d:####\n",
      "{'content': {'analysis_id': 'assay_1'}, 'analysis_type': 'upload', 'data_bundles': [{'bundle_uuid': 'a4aed7f3-7fbf-45ee-b2aa-4402126414ab', 'files': [{'content-type': 'application/gzip; dcp-type=data', 'crc32c': '4ef74578', 'indexed': False, 'name': 'R1.fastq.gz', 's3_etag': 'c7bbee4c46bbf29432862e05830c8f39', 'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3', 'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29', 'size': 125191, 'uuid': '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d', 'version': '2017-12-21T164604.038618Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}, {'content-type': 'application/gzip; dcp-type=data', 'crc32c': '69987b3e', 'indexed': False, 'name': 'R2.fastq.gz', 's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae', 'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21', 'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba', 'size': 130024, 'uuid': 'cedf25a8-ce28-4025-a686-459b220fc825', 'version': '2017-12-21T164605.003656Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}]}]}\n",
      "\n",
      "#########PRINTING ITEM IN __merge_lists###############:\n",
      "{'bundle_uuid': 'a4aed7f3-7fbf-45ee-b2aa-4402126414ab', 'files': [{'content-type': 'application/gzip; dcp-type=data', 'crc32c': '4ef74578', 'indexed': False, 'name': 'R1.fastq.gz', 's3_etag': 'c7bbee4c46bbf29432862e05830c8f39', 'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3', 'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29', 'size': 125191, 'uuid': '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d', 'version': '2017-12-21T164604.038618Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}, {'content-type': 'application/gzip; dcp-type=data', 'crc32c': '69987b3e', 'indexed': False, 'name': 'R2.fastq.gz', 's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae', 'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21', 'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba', 'size': 130024, 'uuid': 'cedf25a8-ce28-4025-a686-459b220fc825', 'version': '2017-12-21T164605.003656Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}]}\n",
      "\n",
      "<class 'dict'>\n",
      "####GET_VALUE d:####\n",
      "{'bundle_uuid': 'a4aed7f3-7fbf-45ee-b2aa-4402126414ab', 'files': [{'content-type': 'application/gzip; dcp-type=data', 'crc32c': '4ef74578', 'indexed': False, 'name': 'R1.fastq.gz', 's3_etag': 'c7bbee4c46bbf29432862e05830c8f39', 'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3', 'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29', 'size': 125191, 'uuid': '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d', 'version': '2017-12-21T164604.038618Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}, {'content-type': 'application/gzip; dcp-type=data', 'crc32c': '69987b3e', 'indexed': False, 'name': 'R2.fastq.gz', 's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae', 'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21', 'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba', 'size': 130024, 'uuid': 'cedf25a8-ce28-4025-a686-459b220fc825', 'version': '2017-12-21T164605.003656Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}]}\n",
      "\n",
      "#########PRINTING ITEM IN __merge_lists###############:\n",
      "{'bundle_uuid': 'a4aed7f3-7fbf-45ee-b2aa-4402126414ab', 'files': [{'content-type': 'application/gzip; dcp-type=data', 'crc32c': '4ef74578', 'indexed': False, 'name': 'R1.fastq.gz', 's3_etag': 'c7bbee4c46bbf29432862e05830c8f39', 'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3', 'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29', 'size': 125191, 'uuid': '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d', 'version': '2017-12-21T164604.038618Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}, {'content-type': 'application/gzip; dcp-type=data', 'crc32c': '69987b3e', 'indexed': False, 'name': 'R2.fastq.gz', 's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae', 'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21', 'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba', 'size': 130024, 'uuid': 'cedf25a8-ce28-4025-a686-459b220fc825', 'version': '2017-12-21T164605.003656Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}]}\n",
      "\n",
      "<class 'dict'>\n",
      "####GET_VALUE d:####\n",
      "{'bundle_uuid': 'a4aed7f3-7fbf-45ee-b2aa-4402126414ab', 'files': [{'content-type': 'application/gzip; dcp-type=data', 'crc32c': '4ef74578', 'indexed': False, 'name': 'R1.fastq.gz', 's3_etag': 'c7bbee4c46bbf29432862e05830c8f39', 'sha1': '17f8b4be0cc6e8281a402bb365b1283b458906a3', 'sha256': 'fe6d4fdfea2ff1df97500dcfe7085ac3abfb760026bff75a34c20fb97a4b2b29', 'size': 125191, 'uuid': '67b3e2eb-62b8-49bd-9e5c-aa8e80a8bd3d', 'version': '2017-12-21T164604.038618Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}, {'content-type': 'application/gzip; dcp-type=data', 'crc32c': '69987b3e', 'indexed': False, 'name': 'R2.fastq.gz', 's3_etag': 'a3a9f23d07cfc5e40a4c3a8adf3903ae', 'sha1': 'f166b6952e30a41e1409e7fb0cb0fb1ad93f3f21', 'sha256': 'c305bee37b3c3735585e11306272b6ab085f04cd22ea8703957b4503488cfeba', 'size': 130024, 'uuid': 'cedf25a8-ce28-4025-a686-459b220fc825', 'version': '2017-12-21T164605.003656Z', 'format': 'fastq.gz', 'bundle_type': 'scRNA-Seq Upload'}]}\n",
      "\n",
      "#########PRINTING ITEM IN __merge_lists###############:\n",
      "{'content': {'core': {'type': 'project', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project.json', 'schema_version': '4.6.1'}, 'description': 'Q3_DEMO-We report transcriptomes from 430 single glioblastoma cells isolated from 5 individual tumors and 102 single cells from gliomasphere cells lines generated using SMART-seq. In addition, we report population RNA-seq from the five tumors as well as RNA-seq from cell lines derived from 3 tumors (MGH26, MGH28, MGH31) cultured under serum free (GSC) and differentiated (DGC) conditions. This dataset highlights intratumoral heterogeneity with regards to the expression of de novo derived transcriptional modules and established subtype classifiers. Overall design: Operative specimens from five glioblastoma patients (MGH26, MGH28, MGH29, MGH30, MGH31) were acutely dissociated, depleted for CD45+ inflammatory cells and then sorted as single cells (576 samples). Population controls for each tumor were isolated by sorting 2000-10000 cells and processed in parallel (5 population control samples). Single cells from two established cell lines, GBM6 and GBM8, were also sorted as single cells (192 samples). SMART-seq protocol was implemented to generate single cell full length transcriptomes (modified from Shalek, et al Nature 2013) and sequenced using 25 bp paired end reads. Single cell cDNA libraries for MGH30 were resequenced using 100 bp paired end reads to allow for isoform and splice junction reconstruction (96 samples, annotated MGH30L). Cells were also cultured in serum free conditions to generate gliomasphere cell lines for MGH26, MGH28, and MGH31 (GSC) which were then differentiated using 10% serum (DGC). Population RNA-seq was performed on these samples (3 GSC, 3 DGC, 6 total). The initial dataset included 875 RNA-seq libraries (576 single glioblastoma cells, 96 resequenced MGH30L, 192 single gliomasphere cells, 5 tumor population controls, 6 population libraries from GSC and DGC samples). Data was processed as described below using RSEM for quantification of gene expression. 5,948 genes with the highest composite expression either across all single cells combined (average log2(TPM)>4.5) or within a single tumor (average log2(TPM)>6 in at least one tumor) were included. Cells expressing less than 2,000 of these 5,948 genes were excluded. The final processed dataset then included 430 primary single cell glioblastoma transcriptomes, 102 single cell transcriptomes from cell lines(GBM6,GBM8), 5 population controls (1 for each tumor), and 6 population libraries from cell lines derived from the tumors (GSC and DGC for MGH26, MGH28 and MGH31). The final matrix (GBM_data_matrix.txt) therefore contains 5948 rows (genes) quantified in 543 samples (columns). Please note that the samples which are not included in the data processing are indicated in the sample description field.', 'contributors': [{'name': 'Q3_DEMO-MintTeam', 'email': 'dummy@email.com'}], 'submitters': [{'name': 'Q3_DEMO-MintTeam', 'email': 'dummy@email.com'}], 'publications': [], 'project_id': 'Q3_DEMO-project_PRJNA248302', 'name': 'Q3_DEMO-Single cell RNA-seq of primary human glioblastomas'}, 'core': {'type': 'project_bundle', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project_bundle.json', 'schema_version': '4.6.1'}, 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.684Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:21.203Z', 'document_id': 'abc0cd74-45ce-4199-bc25-734cf09fec13', 'user': 'anonymousUser'}}\n",
      "\n",
      "<class 'dict'>\n",
      "####GET_VALUE d:####\n",
      "{'content': {'core': {'type': 'project', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project.json', 'schema_version': '4.6.1'}, 'description': 'Q3_DEMO-We report transcriptomes from 430 single glioblastoma cells isolated from 5 individual tumors and 102 single cells from gliomasphere cells lines generated using SMART-seq. In addition, we report population RNA-seq from the five tumors as well as RNA-seq from cell lines derived from 3 tumors (MGH26, MGH28, MGH31) cultured under serum free (GSC) and differentiated (DGC) conditions. This dataset highlights intratumoral heterogeneity with regards to the expression of de novo derived transcriptional modules and established subtype classifiers. Overall design: Operative specimens from five glioblastoma patients (MGH26, MGH28, MGH29, MGH30, MGH31) were acutely dissociated, depleted for CD45+ inflammatory cells and then sorted as single cells (576 samples). Population controls for each tumor were isolated by sorting 2000-10000 cells and processed in parallel (5 population control samples). Single cells from two established cell lines, GBM6 and GBM8, were also sorted as single cells (192 samples). SMART-seq protocol was implemented to generate single cell full length transcriptomes (modified from Shalek, et al Nature 2013) and sequenced using 25 bp paired end reads. Single cell cDNA libraries for MGH30 were resequenced using 100 bp paired end reads to allow for isoform and splice junction reconstruction (96 samples, annotated MGH30L). Cells were also cultured in serum free conditions to generate gliomasphere cell lines for MGH26, MGH28, and MGH31 (GSC) which were then differentiated using 10% serum (DGC). Population RNA-seq was performed on these samples (3 GSC, 3 DGC, 6 total). The initial dataset included 875 RNA-seq libraries (576 single glioblastoma cells, 96 resequenced MGH30L, 192 single gliomasphere cells, 5 tumor population controls, 6 population libraries from GSC and DGC samples). Data was processed as described below using RSEM for quantification of gene expression. 5,948 genes with the highest composite expression either across all single cells combined (average log2(TPM)>4.5) or within a single tumor (average log2(TPM)>6 in at least one tumor) were included. Cells expressing less than 2,000 of these 5,948 genes were excluded. The final processed dataset then included 430 primary single cell glioblastoma transcriptomes, 102 single cell transcriptomes from cell lines(GBM6,GBM8), 5 population controls (1 for each tumor), and 6 population libraries from cell lines derived from the tumors (GSC and DGC for MGH26, MGH28 and MGH31). The final matrix (GBM_data_matrix.txt) therefore contains 5948 rows (genes) quantified in 543 samples (columns). Please note that the samples which are not included in the data processing are indicated in the sample description field.', 'contributors': [{'name': 'Q3_DEMO-MintTeam', 'email': 'dummy@email.com'}], 'submitters': [{'name': 'Q3_DEMO-MintTeam', 'email': 'dummy@email.com'}], 'publications': [], 'project_id': 'Q3_DEMO-project_PRJNA248302', 'name': 'Q3_DEMO-Single cell RNA-seq of primary human glioblastomas'}, 'core': {'type': 'project_bundle', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project_bundle.json', 'schema_version': '4.6.1'}, 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.684Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:21.203Z', 'document_id': 'abc0cd74-45ce-4199-bc25-734cf09fec13', 'user': 'anonymousUser'}}\n",
      "\n",
      "#########PRINTING ITEM IN __merge_lists###############:\n",
      "{'content': {'core': {'type': 'project', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project.json', 'schema_version': '4.6.1'}, 'description': 'Q3_DEMO-We report transcriptomes from 430 single glioblastoma cells isolated from 5 individual tumors and 102 single cells from gliomasphere cells lines generated using SMART-seq. In addition, we report population RNA-seq from the five tumors as well as RNA-seq from cell lines derived from 3 tumors (MGH26, MGH28, MGH31) cultured under serum free (GSC) and differentiated (DGC) conditions. This dataset highlights intratumoral heterogeneity with regards to the expression of de novo derived transcriptional modules and established subtype classifiers. Overall design: Operative specimens from five glioblastoma patients (MGH26, MGH28, MGH29, MGH30, MGH31) were acutely dissociated, depleted for CD45+ inflammatory cells and then sorted as single cells (576 samples). Population controls for each tumor were isolated by sorting 2000-10000 cells and processed in parallel (5 population control samples). Single cells from two established cell lines, GBM6 and GBM8, were also sorted as single cells (192 samples). SMART-seq protocol was implemented to generate single cell full length transcriptomes (modified from Shalek, et al Nature 2013) and sequenced using 25 bp paired end reads. Single cell cDNA libraries for MGH30 were resequenced using 100 bp paired end reads to allow for isoform and splice junction reconstruction (96 samples, annotated MGH30L). Cells were also cultured in serum free conditions to generate gliomasphere cell lines for MGH26, MGH28, and MGH31 (GSC) which were then differentiated using 10% serum (DGC). Population RNA-seq was performed on these samples (3 GSC, 3 DGC, 6 total). The initial dataset included 875 RNA-seq libraries (576 single glioblastoma cells, 96 resequenced MGH30L, 192 single gliomasphere cells, 5 tumor population controls, 6 population libraries from GSC and DGC samples). Data was processed as described below using RSEM for quantification of gene expression. 5,948 genes with the highest composite expression either across all single cells combined (average log2(TPM)>4.5) or within a single tumor (average log2(TPM)>6 in at least one tumor) were included. Cells expressing less than 2,000 of these 5,948 genes were excluded. The final processed dataset then included 430 primary single cell glioblastoma transcriptomes, 102 single cell transcriptomes from cell lines(GBM6,GBM8), 5 population controls (1 for each tumor), and 6 population libraries from cell lines derived from the tumors (GSC and DGC for MGH26, MGH28 and MGH31). The final matrix (GBM_data_matrix.txt) therefore contains 5948 rows (genes) quantified in 543 samples (columns). Please note that the samples which are not included in the data processing are indicated in the sample description field.', 'contributors': [{'name': 'Q3_DEMO-MintTeam', 'email': 'dummy@email.com'}], 'submitters': [{'name': 'Q3_DEMO-MintTeam', 'email': 'dummy@email.com'}], 'publications': [], 'project_id': 'Q3_DEMO-project_PRJNA248302', 'name': 'Q3_DEMO-Single cell RNA-seq of primary human glioblastomas'}, 'core': {'type': 'project_bundle', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project_bundle.json', 'schema_version': '4.6.1'}, 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.684Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:21.203Z', 'document_id': 'abc0cd74-45ce-4199-bc25-734cf09fec13', 'user': 'anonymousUser'}}\n",
      "\n",
      "<class 'dict'>\n",
      "####GET_VALUE d:####\n",
      "{'content': {'core': {'type': 'project', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project.json', 'schema_version': '4.6.1'}, 'description': 'Q3_DEMO-We report transcriptomes from 430 single glioblastoma cells isolated from 5 individual tumors and 102 single cells from gliomasphere cells lines generated using SMART-seq. In addition, we report population RNA-seq from the five tumors as well as RNA-seq from cell lines derived from 3 tumors (MGH26, MGH28, MGH31) cultured under serum free (GSC) and differentiated (DGC) conditions. This dataset highlights intratumoral heterogeneity with regards to the expression of de novo derived transcriptional modules and established subtype classifiers. Overall design: Operative specimens from five glioblastoma patients (MGH26, MGH28, MGH29, MGH30, MGH31) were acutely dissociated, depleted for CD45+ inflammatory cells and then sorted as single cells (576 samples). Population controls for each tumor were isolated by sorting 2000-10000 cells and processed in parallel (5 population control samples). Single cells from two established cell lines, GBM6 and GBM8, were also sorted as single cells (192 samples). SMART-seq protocol was implemented to generate single cell full length transcriptomes (modified from Shalek, et al Nature 2013) and sequenced using 25 bp paired end reads. Single cell cDNA libraries for MGH30 were resequenced using 100 bp paired end reads to allow for isoform and splice junction reconstruction (96 samples, annotated MGH30L). Cells were also cultured in serum free conditions to generate gliomasphere cell lines for MGH26, MGH28, and MGH31 (GSC) which were then differentiated using 10% serum (DGC). Population RNA-seq was performed on these samples (3 GSC, 3 DGC, 6 total). The initial dataset included 875 RNA-seq libraries (576 single glioblastoma cells, 96 resequenced MGH30L, 192 single gliomasphere cells, 5 tumor population controls, 6 population libraries from GSC and DGC samples). Data was processed as described below using RSEM for quantification of gene expression. 5,948 genes with the highest composite expression either across all single cells combined (average log2(TPM)>4.5) or within a single tumor (average log2(TPM)>6 in at least one tumor) were included. Cells expressing less than 2,000 of these 5,948 genes were excluded. The final processed dataset then included 430 primary single cell glioblastoma transcriptomes, 102 single cell transcriptomes from cell lines(GBM6,GBM8), 5 population controls (1 for each tumor), and 6 population libraries from cell lines derived from the tumors (GSC and DGC for MGH26, MGH28 and MGH31). The final matrix (GBM_data_matrix.txt) therefore contains 5948 rows (genes) quantified in 543 samples (columns). Please note that the samples which are not included in the data processing are indicated in the sample description field.', 'contributors': [{'name': 'Q3_DEMO-MintTeam', 'email': 'dummy@email.com'}], 'submitters': [{'name': 'Q3_DEMO-MintTeam', 'email': 'dummy@email.com'}], 'publications': [], 'project_id': 'Q3_DEMO-project_PRJNA248302', 'name': 'Q3_DEMO-Single cell RNA-seq of primary human glioblastomas'}, 'core': {'type': 'project_bundle', 'schema_url': 'https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.1/json_schema/project_bundle.json', 'schema_version': '4.6.1'}, 'hca_ingest': {'accession': '', 'submissionDate': '2017-12-21T16:44:39.684Z', 'lastModifiedUser': 'anonymousUser', 'updateDate': '2017-12-21T16:45:21.203Z', 'document_id': 'abc0cd74-45ce-4199-bc25-734cf09fec13', 'user': 'anonymousUser'}}\n",
      "\n",
      "INDEXING DONE\n"
     ]
    }
   ],
   "source": [
    "# Define helper method to open files\n",
    "def open_and_return_json(file_path):\n",
    "    \"\"\"\n",
    "    Opens and returns the contents of the json file given in file_path\n",
    "    :param file_path: Path of a json file to be opened\n",
    "    :return: Returns an obj with the contents of the json file\n",
    "    \"\"\"\n",
    "    with open(file_path, 'r') as file_:\n",
    "        loaded_file = json.load(file_)\n",
    "    return loaded_file\n",
    "\n",
    "# Get the index's settings\n",
    "index_settings = open_and_return_json('chalicelib/settings.json')\n",
    "# Get the index overall config\n",
    "index_mapping_config = open_and_return_json('chalicelib/config.json')\n",
    "\n",
    "file_indexer = FileIndexer(metadata_files,\n",
    "                           data_files,\n",
    "                           es,\n",
    "                           \"file_index_v4\",\n",
    "                           \"doc\",\n",
    "                           index_settings=index_settings,\n",
    "                           index_mapping_config=index_mapping_config)\n",
    "\n",
    "assay_indexer = AssayOrientedIndexer(metadata_files,\n",
    "                                     data_files,\n",
    "                                     es,\n",
    "                                     \"assay_index_v4\",\n",
    "                                     \"doc\",\n",
    "                                     index_settings=index_settings,\n",
    "                                     index_mapping_config=index_mapping_config)\n",
    "\n",
    "file_indexer.index(bundle_uuid, bundle_version)\n",
    "assay_indexer.index(bundle_uuid, bundle_version)\n",
    "print(\"INDEXING DONE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
